---
title: "Jeager Engram Cell Gene signatures in other datasets"
author: "Angus Campbell"
date: "11/02/2022"
output: github_document
---

```{r include = FALSE}
#get libraries
setwd("C:/Users/angus/Desktop/PavLabEngrams/EngramCellClassifier")

library(tidyverse)
library(GEOquery)
library(AnnotationDbi)
library(randomForest)
library(data.table)
library(reshape2)
library(FactoMineR)
library(factoextra)
library(Rtsne)
library(dplyr)
library(Seurat) #installed with remotes
library(stringr)
library(patchwork)
library(metap)
library(cqn)
library(GeoTcgaData)
library(ggplot2)
library(tictoc)
library(pheatmap)
library(SeuratDisk) # installed with remotes
#reticulate::py_install(packages ='umap-learn')

#Seurat reference manual: https://cloud.r-project.org/web/packages/Seurat/Seurat.pdf
```


Introduction
---

Here I have used supplemental data 4 from Jeager et al., (2018) which is a list of the DEGs from Fos+ dentate granule cells (DGCs) at various time points vs Fos- DGCs at various time points.  The authors assigned these DEGs to qualitatively similar time courses based on thier status as DEGs across 3 different time points (1hr, 4hrs, and 5hrs).  While not rigorous this manual assissignment is still interesting to use, perhaps later we can redo this with a more rigorous exploration of the time course data.  This first experiemnt was all female mice.  Regardless I use this DEG list to cluster DGCs from the first experiment of Hochgerner et al., (2018).  This second dataset of ~5k cells was sampled from male and female mice across development.  The sample of interest for me were samples from post natal day 37 which is relatively close to the 7-8.75 week old mice (in days they are ~P49-P62) in Jeager.  Though not an ideal match it is better than the other dataset which contains 24k cells but range from embryonic mice to post natal day 120 and 132.  Jeager et al. (2018) was done entirely using female mice while Hochgerner was done with a mixture of males and females.

The Hochgerner dataset contains some peculiarities.  They have pooled cells between males and females to reduce batch effects, but its difficult to guess which cells came from which.  Hochgerner et al., (2018)  assigned sex to cells based on number of male vs female genes in each cell.  They provide metadata about the number fo males and females pooled for the batch each cell was drawn from but in the post natal days 24 and 35 in the Hoch5k data there are some that simply cannot be known.


### Jeager et al., (2018) meta data and counts

Load Jeager data, some preprocessing is required.

```{r}
#Jeagers meta rows are a little out of order wrt their counts, i.e. rows do no correspond to cells order we fix that in a bit
jeager2018_counts <- bind_cols(read.table('Jeager2018_GSE98679/GSE98679_count.txt.gz', header = TRUE, check.names = FALSE),
                               read.table('Jeager2018_GSE98679/GSE98679_v2_GSM3308862-GSM3309413_count.txt.gz', header = TRUE, check.names = FALSE))

jeager2018_meta <- read.csv('Jeager2018_GSE98679/SraRunTable.txt', header = TRUE)
jeager2018_meta = jeager2018_meta[c(1:46,599:912,47:598),] #here we fix the order
rownames(jeager2018_meta) <- c(1:912)

jeager2018_meta$predicted_cell_type <- as.character(lapply(jeager2018_meta$predicted_cell_type, function(x) if (x=="") {"DG"} else {x}))
jeager2018_meta$predicted_cell_type <- lapply(jeager2018_meta$predicted_cell_type, function(x) if (x=="") {"DG"} else {x})

jeager2018_meta$fos_status <- as.factor(sapply(as.character(jeager2018_meta$source_name), function(y) if (grepl("_F_", y, fixed=TRUE)) "Fos+" else "Fos-"  ))
jeager2018_meta$fos_status[361:dim(jeager2018_meta)[1]] <- "Fos+"

jeager2018_meta$Mouse_Number[c(361:912)] <- jeager2018_meta$mousingle_number[c(361:912)]
jeager2018_meta <- jeager2018_meta %>% 
  dplyr::select(-mousingle_number)

#filtering out cells as per instructions in Jeager et al., 2018)
under4k <- sapply(jeager2018_counts, function(y) sum(length(which(y>0))))

filtered.idx <- as.numeric(which(under4k>4000))
filtered.idx <- order(c(filtered.idx,194))

jeager2018_counts <-jeager2018_counts[,filtered.idx]
jeager2018_meta <- jeager2018_meta[filtered.idx,]
```



### Getting a handle on realsitic Fos levels in Jeager

<br><br>
Quick check for HC DGCs with fos labelleing.
<br>
```{r}
which((jeager2018_meta$exposure=="HC")&(jeager2018_meta$fos_status=="Fos+"))
```
<br><br>
Just to illustrate what we should expect to see in a population of engram cells we will look at the fos expression across the cells in Jeager based on their labelling strategy.  They used both Fos and Arc as labels as Fos expression is quite transient where as Arc can be stained for with anitibodies for up to 5 hours after the expossure.  Not all the cells antibody labelled are actively transcribing the gene.
<br><br>
```{r}
jeager.fos.idx <- which(rownames(jeager2018_counts)=="Fos")
fosneg.HC <- which((jeager2018_meta$exposure=="HC")&(jeager2018_meta$fos_status=="Fos-"))

hist(as.numeric(jeager2018_counts[jeager.fos.idx, fosneg.HC]), 
     main = "Homecage Fos-AntiBd Negative nuclear Fos read distribution",
     xlab = "Fos Read Count")


```
<br>
```{r}
fosneg.1hr <- which((jeager2018_meta$exposure=="1hr")&(jeager2018_meta$fos_status=="Fos-"))


hist(as.numeric(jeager2018_counts[jeager.fos.idx, fosneg.1hr]), 
     main = "1 Hour pst NE Fos-AntiBd Negative nuclear Fos read distribution",
     xlab = "Fos Read Count")
```

```{r}
fospos.1hr <- which((jeager2018_meta$exposure=="1hr")&(jeager2018_meta$fos_status=="Fos+"))

hist(as.numeric(jeager2018_counts[jeager.fos.idx, fospos.1hr]),
     main = "1Hour post-NE Fos-AntiBd Positive nuclear Fos read distribution",
     xlab = "Fos Read Count") 
```


```{r}
#making ggplot histpgram 
fnHC <- as.numeric(jeager2018_counts[jeager.fos.idx, fosneg.HC ])
fn1hr <- as.numeric(jeager2018_counts[jeager.fos.idx, fosneg.1hr])
fp1hr <- as.numeric(jeager2018_counts[jeager.fos.idx, fospos.1hr])

df <- data.frame( fos = as.numeric(fnHC), 
                  cond = factor(rep("Fos- HC", length(fnHC)))
                  )

df2 <- data.frame( fos = as.numeric(fn1hr), 
                  cond = factor(rep("Fos- 1hr", length(fn1hr)))
                  )

df3 <-  data.frame( fos = as.numeric(fp1hr), 
                  cond = factor(rep("Fos+ 1hr", length(fp1hr)))
                  )

df <- rbind(df,df2,df3)

p <- ggplot(df, aes(x=fos, color = cond)) + 
  geom_histogram(fill =  "white") +
  xlab("Fos Reads (seq. depth of 1.5 milion reads per cell)") +
  ylab("Frequency") +
  theme_classic() + 
  theme(legend.title = element_blank()) +
  ggtitle("Freqency of Fos Counts in Jeager Groups")+
  theme(plot.title = element_text(hjust = 0.5))+
  theme(axis.text = element_text(face = "bold")) +
  scale_x_discrete(limits = c(0, 6000)) +
  scale_y_continuous(limits = c(0, 350))

p
```

<br>
So in general there are a few cells with a low level expression of fos, and then this small group of outliers in with a high fos count.  In general there are less Fos+ (by antibody staining) cells than FOs- and they are not all transcribing Fos, though a greater proportion of cells in the Fos+ group are actively transcribing it.  This is in line with the sparse coding of events observed in the brain.  It is notable that within the group that was deemed to be an "engram cell" via antibody staining for fos that there is a not insignificant fraction of the fos positive cells which are not expressing fos.  This could be due to off target effects, the authors themselves make note that in CA1 cells they had stained for fos or arc did not express those markers.  It would be better to have a true engram label dataset to work with like Chen et al., (2018) but this is what we had for hippocampus so it will have to do.  Jeager's sequencing was much deeper than Hochgerner's (1.5 million avg reads per cell to ~43k avg reads per cell) and Arc is quite lowly expressed so its difficult to detect in the Hochgerner data.
<br>
<br>
For our purposes of computationally identify engram cells we want to identify gene's that are robustly expressed for long periods of time as well as gene's that are robustly expressed within particular contexts like at certain time points.  This is why it's better to use a set of DEGs from a confirmed engram experiment rather than looking individually for Erg1 or Inhba or Arc or Fos etc etc.  But in general at least for recently active cells you would expect them to be rare cells, over expressing DEGs.
<br>
<br>

Here we could also discuss the issues arrising from what kind of cells we are looking for.  Add images of the possible sources of IEGs you are seeing.


Looking for Engram labelling gene's in Hochgerner et al., (2018) at P35
---
<br>
<br>
This section is basically just copy pasta from [Seurats tutorials](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html). Here I am using the DEGs published by Jeager et al., (2018).  First we will load the data and the engram cell DEG list, refered to as Jeager DEGs or Jeager Engram DEGs hereafter.  We will stick to the adult DGCs 
<br>
<br>
```{r}
#loading the 5k dataset
# for now I think it is best to keep it seperate from the 24k data as it has only half the genes
# of the 
testsetpath <- "C:/Users/angus/Desktop/test_datasets"

hochgerner5k_2018_counts <- read.table(paste(testsetpath,"/Hochgerner2018/GSE95315_10X_expression_data_v2.tab.gz", sep=""))

colnames(hochgerner5k_2018_counts) <- hochgerner5k_2018_counts[1,]
rownames(hochgerner5k_2018_counts) <- hochgerner5k_2018_counts[,1]

hochgerner5k_2018_meta <- hochgerner5k_2018_counts %>% 
  dplyr::slice(c(1:3)) %>%
  t() %>%
  data.frame %>%
  dplyr::slice(-1) %>% 
  dplyr::select(-cellid)

hochgerner5k_2018_counts <- hochgerner5k_2018_counts %>% 
  dplyr::select(-cellid) %>% 
  dplyr::slice(-c(1:3))


#Load Jeagers DEGs and Check which genes are shared

Jeager.DEGs <- read.csv("Jeager2018_GSE98679/jeager_DEGs_grouped.csv", header = T)

colnames(Jeager.DEGs)[1] <- "GeneID"

DEGs.idx <- Jeager.DEGs$GeneID

DEGs.idx <- DEGs.idx[(DEGs.idx %in% rownames(hochgerner5k_2018_counts))& (DEGs.idx %in% rownames(jeager2018_counts))]

Jeager.DEGs <- Jeager.DEGs[Jeager.DEGs$GeneID %in% DEGs.idx,]

Jeager.DEGs$hoch5k.idx <- which(rownames(hochgerner5k_2018_counts) %in% Jeager.DEGs$GeneID)
```


###  Full gene set in hochgerner
<br>
<br>
First we will cluster within the adult DGCs at post natal day 35 to see if engram cells are aparent just from that.  Although one would expect that some previous group would have noted their existance just from looking for simple cell types.  Previous work on another dataset of DGCs reports no such cell type distinguished by high IEG expression (Erwin et al., 2020; Habib et al., 2016).


```{r}
#get index locations of these gene's in the other datasets
hoch5k.GC_Adult.p35.idx <- (hochgerner5k_2018_meta$age.days.=="35") | (hochgerner5k_2018_meta$age.days.=="35*")
hoch5k.GC_Adult.p35.idx <- (hoch5k.GC_Adult.p35.idx) & (hochgerner5k_2018_meta$cluster_name == "Granule-mature")
hoch5k.GC_Adult.p35.idx <- which(hoch5k.GC_Adult.p35.idx)


#create seurat object just for the Hoch 5k data, with only the JEager DEG counts
#Hoch5k.GCadult <- CreateSeuratObject(counts = hochgerner5k_2018_counts[Jeager.DEGs$hoch5k.idx, hoch5k.GC_Adult.p35.idx]) 

Hoch5k.GCadult <- CreateSeuratObject(counts = hochgerner5k_2018_counts[, hoch5k.GC_Adult.p35.idx])
#this is unesscesary we can do this by changing the, 
#well on second thoguth may be we do need to do this because noramlization may be important to be constrained to just the IEGs
#I think seurat borrows information from other genes. test this
Hoch5k.GCadult
```


### QUALITY CONTROL

Cells and Gene counts
  The dataset provided has already been filtered for cells which, the dataset I refer to as hochgerner 5k is here called dataset A.  It's accession number is GSE95315.  Of the original 6090 cells collected in dataset A they were wilted  down to the 5454 cells in the dataset according to the following measures: "10x Chromium Dataset A (Dataset C) data files were loaded and merged to one dataset. For more uniform representation of timepoints, Dataset C data files were downsampled as indicated in Supplementary Table 5 (‘Cells_analysed’). Valid cells were defined as having >600 (C:>800) total genes, between 800 and 20,000 (C: 1,000–30,000) molecules, and a ratio of molecules to genes >1.2. This resulted in 6,090 (C: 25,546) cells. Next, we removed 46 (C: 732) doublet cells based on coexpression (>1 molecule) of any pair of the following marker genes: Stmn2 (neurons), Mog (oligodendrocytes), Aldoc (astrocytes), C1qc (microglia), and Cldn5 (endothelial). "
 
We can check this by plotting the gene counts and the mitochondria RNA, it is clear that the cells in the dataset have already been subject to the above selection process.  Additionally we will check for over expression of mitochondrial genes as is suggested by the seurate documentation.

```{r}
#make a seurat object of the adult granule cells
Hoch5k.GCadult[["percent.mt"]] <- PercentageFeatureSet(Hoch5k.GCadult, pattern = "^MT-")
```


```{r}
VlnPlot(Hoch5k.GCadult, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


```{r}
plot1 <- FeatureScatter(Hoch5k.GCadult, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Hoch5k.GCadult, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

```{r}

# This is for the filerting out those with features that 
# I'm not sure what the appropriate number for us is, in the tutorial there were 13700 nfeatures
#and the filterd out cells with less then 200 and greater than 2500, so I guess  cells with more than
# a quarter of nFeatures and less than 10% of that is unacceptable?  
Hoch5k.GCadult <- subset(Hoch5k.GCadult, subset = nFeature_RNA > 600 & nFeature_RNA < 20000 & percent.mt < 5)

#note I am going to skip some QC steps here, but we do need to do that in the future
Hoch5k.GCadult <- NormalizeData(Hoch5k.GCadult) # Seurat normalization has many steps see publication
# https://www.frontiersin.org/articles/10.3389/fgene.2020.00041/full describes and compares it to others
Hoch5k.GCadult <- FindVariableFeatures(Hoch5k.GCadult) # Identifies features that are outliers on a ’mean variability plot’.

```

Now we can check the most variable genes which seurat will restrict further analyses to.  Two of the top 10 most variable genes are in the Jeager DEGs, Inhba and Tax1bp1.  

```{r}
top10 <-  head(VariableFeatures(Hoch5k.GCadult), 10)
plot1 <- VariableFeaturePlot(Hoch5k.GCadult)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE) + NoLegend()
plot2 + ggtitle("Seurat Most Variable Genes Hochgerner")
```

Lets look at the top 100 of the 2000 genes seurat will use and check the timepoint(s) Jeager et al., (2018) assigned them to.  For the most part these DEGs are positively expressed.
```{r}
top100 <-  head(VariableFeatures(Hoch5k.GCadult), 100)
top100[which(top100 %in% Jeager.DEGs$GeneID)]
Jeager.DEGs$group.number[Jeager.DEGs$GeneID %in% top100]
Jeager.DEGs$group[Jeager.DEGs$GeneID %in% top100]
```

```{r}
#scaling for dimnesionality reduction
all.genes <- rownames(Hoch5k.GCadult)
Hoch5k.GCadult <- ScaleData(Hoch5k.GCadult, features = all.genes) 

Hoch5k.GCadult <- RunPCA(Hoch5k.GCadult , features = VariableFeatures(object = Hoch5k.GCadult),
                         npcs = 20) 
```
<br>
<br>


```{r}
#iop
print(Hoch5k.GCadult[["pca"]], dims = 1:6, nfeatures = 10)

```



Now we will compute and plot the JackStraw score.  
```{r}
#Dsicovering dimensionality of these activity markers
Hoch5k.GCadult <- JackStraw(Hoch5k.GCadult, num.replicate = 100)
Hoch5k.GCadult <- ScoreJackStraw(Hoch5k.GCadult, dims = 1:20)
```
<br>
<br>
From Seurat's documentation 
https://satijalab.org/seurat/reference/scorejackstraw
"Plots the results of the JackStraw analysis for PCA significance. For each PC, plots a QQ-plot comparing the distribution of p-values for all genes across each PC, compared with a uniform distribution. Also determines a p-value for the overall significance of each PC (see Details)."
<br>
https://satijalab.org/seurat/reference/jackstrawplot#details
"Significant PCs should show a p-value distribution (black curve) that is strongly skewed to the left compared to the null distribution (dashed line) The p-value for each PC is based on a proportion test comparing the number of genes with a p-value below a particular threshold (score.thresh), compared with the proportion of genes expected under a uniform distribution of p-values"
<br>
<br>

```{r}
JackStrawPlot(Hoch5k.GCadult, dims = 1:20)
```
<br>
<br>
Up to Pc3 we see decently significance, though later we get 4 cluster, though tow of them are barely distinguishable and posses basically no activity dependent markers (Jeager DEGs whihc could serve as markers.
<br>
<br>

```{r}
p <- ElbowPlot(Hoch5k.GCadult)
p+ ggtitle("Explained Varience of Hochgerner PCs") + theme(plot.title = element_text(hjust = 0.5))
```
<br>
<br>
Shows more or less the same thing, 4 isn't quite at the elbow but it may as well be.  I played with the QC thresholds and if you increase the number of max nFeatures to high you will reduce to a single PC.
<br>
<br>


### Number of Dimensions affects results

<br>
<br>
I would say that the best cut off would be to go up to dimensions 8, but when this is chosen (by setting dims = 1:8 in FindNeighbors) that doesn't 

```{r}
#Adding metadata to identify the 
Hoch5k.GCadult <- AddMetaData(object = Hoch5k.GCadult, 
                               metadata = hochgerner5k_2018_meta[ hoch5k.GC_Adult.p35.idx, ])

#Clustering, we will visualize in UMP later
Hoch5k.GCadult <- FindNeighbors(Hoch5k.GCadult, dims = 1:3)
Hoch5k.GCadult <- FindClusters(Hoch5k.GCadult, resolution = 0.2, algorithm = 1)
```


```{r}
Hoch5k.GCadult <- RunUMAP(Hoch5k.GCadult, dims = 1:4)

DimPlot(Hoch5k.GCadult, reduction ="umap")
```

```{r}
DimPlot(Hoch5k.GCadult, reduction ="pca")
```



```{r}
Hoch5k.GCadult.markers.3dims <- FindAllMarkers(Hoch5k.GCadult, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
seurat_clusters.3dims <- Hoch5k.GCadult@meta.data$seurat_clusters
```
```{r}
for (i in c(0:2)){
  print(paste("Cluster ", as.character(i)))
  mrks <- Hoch5k.GCadult.markers.3dims$gene[Hoch5k.GCadult.markers.3dims$cluster==i & Hoch5k.GCadult.markers.3dims$p_val_adj<0.05]
  print(mrks)
  print("Gene in the Jeager DEGs are....")
  print(Jeager.DEGs$GeneID[which(Jeager.DEGs$GeneID %in% mrks)])
  print("And their respective jeager goupings are...")
  print(Jeager.DEGs$group[which(Jeager.DEGs$GeneID %in% mrks)])
  print("")
  print("")
}

```
```{r}
Hoch5k.GCadult.markers.3dims %>%
    group_by(cluster) %>%
    slice_max(n = 5, order_by = avg_log2FC)
```


Looking for any non Jeager DEGs markers.

```{r}
Hoch5k.GCadult.markers.3dims %>%
  filter(!(gene %in% Jeager.DEGs$GeneID)&(p_val_adj<0.05))  %>%
  group_by(cluster)  %>%
  slice_max(n = 5, order_by = avg_log2FC)

```
```{r}
#according to Erwin et al., (2020) Cck and Penk should be mutually exclusively expressed
FeaturePlot(Hoch5k.GCadult, features = c("Penk", "Cck"), blend =TRUE)
```

```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Malat1", "Penk"), blend =TRUE)
```


```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Malat1", "Cck"), blend =TRUE)
```

### Run it again with 6 dims

```{r}
# trying again with more dimensions
#Clustering, we will visualize in UMP later
Hoch5k.GCadult <- FindNeighbors(Hoch5k.GCadult, dims = 1:6)
Hoch5k.GCadult <- FindClusters(Hoch5k.GCadult, resolution = 0.2, algorithm = 1)
```


```{r}
Hoch5k.GCadult <- RunUMAP(Hoch5k.GCadult, dims = 1:4)

DimPlot(Hoch5k.GCadult, reduction ="umap")
```

```{r}
p <- DimPlot(Hoch5k.GCadult, reduction ="pca")
p + ggtitle("Hochgerner DGC Subtypes") + theme(plot.title = element_text(hjust = 0.5))

```



```{r}
Hoch5k.GCadult.markers.6dims <- FindAllMarkers(Hoch5k.GCadult, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
seurat_clusters.6dims <- Hoch5k.GCadult@meta.data$seurat_clusters
```
```{r}
for (i in c(0:1)){
  print(paste("Cluster ", as.character(i)))
  mrks <- Hoch5k.GCadult.markers.6dims$gene[Hoch5k.GCadult.markers.6dims$cluster==i & Hoch5k.GCadult.markers.6dims$p_val_adj<0.05]
  print(mrks)
  print("Gene in the Jeager DEGs are....")
  print(Jeager.DEGs$GeneID[which(Jeager.DEGs$GeneID %in% mrks)])
  print("And their respective jeager goupings are...")
  print(Jeager.DEGs$group[which(Jeager.DEGs$GeneID %in% mrks)])
  print("")
  print("")
}

```
```{r}
Hoch5k.GCadult.markers.6dims %>%
    group_by(cluster) %>%
    slice_max(n = 5, order_by = avg_log2FC)
```


```{r}
Hoch5k.GCadult.markers.6dims %>%
  filter(!(gene %in% Jeager.DEGs$GeneID)&(p_val_adj<0.05))  %>%
  group_by(cluster)  %>%
  slice_max(n = 5, order_by = avg_log2FC)

```



```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Fos", "Inhba"))
```
```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Npas4", "Arc"))
```

```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Egr1", "Nrp1"))
```
```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Penk", "Rgs4"), blend = TRUE)
```

So I am observing the Penk and Cck seperation that has been described previously (Erwin et al., 2020).  Overall I am not too sure what to make of this analysis.  As it is within a single cell type it may not make sense to try to cluster it further as continuous variation within these cells may be more accurate (Cembrowski & Mannon, 2018). Because there was a sharp and sudden drop in explained variance between PC3 and PC4 so I reduced the clustering to run using only the top 3 PCs as well as the clustering using up to PC7.  If you include up to PC4 you will find the same seperation of the population into 2 clusters, but neither group really has a remarkably higher presence of Jeager DEGs in it.  I thought that prehaps the other dimesnions may be obscuring some more important processes represented by the top 3 PCs.  Running it including only 3 dimensions gives 4 clusters though I question the validity of the first 2.  

It is worth noting the components of PC3 however.

```{r}
p <- VizDimLoadings(Hoch5k.GCadult, dims = 1, reduction = "pca")
p + ggtitle("Reccuring Component in Full Hochgerner Data") + theme(plot.title = element_text(hjust = 0.5))
```
I would like to point to the presence of Cck as a positive loading and Penk as a negative loading. In Erwin et al., (2020) they
describe a population of mature dentate gyrus cells that are Penk expressing and over represented in the activity labelled cells across a wide variety of hippocampal dependent behaviours during initial training.  Critically in the negative components several other Jeager DEGs including Rgs4 and Inhba, in fact Rgs4 is a marker for cluster 1 in the 7 dimensional clustering.  Cck and Penk cells did not overlap, nor do they here as shown in the figure below.  Penk expressing cells are also rare as would be expected.

```{r}
#according to Erwin et al., (2020) Cck and Penk should be mutually exclusively expressed
FeaturePlot(Hoch5k.GCadult, features = c("Penk", "Cck"))
```

```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Malat1", "Cck"))
```
It appears that this long non-coding RNA Malat1 which is known to be involved in synapse formation is also expressed in opposition to Cck, in fact it determines one poll of the first principle component.  Should try controlling for spatial gradient as well ask mark how to do this.  

```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Malat1", "Penk"), blend = TRUE)
```

So even though Penk does not show up as a marker, and Malat1 is lowly expressed throughout, very high Malat1 expressing is uncommon, and does not overlap with Cck nor Penk.  Likewise Penk does not overlap with CcK.  Even though our clustering algirthms do not turn these up I believe these may be cases of stable gene expression.


```{r}
# > which(rownames(hochgerner5k_2018_counts)=="Penk")
# [1] 9198
# 
# > which(rownames(hochgerner5k_2018_counts)=="Cck")
# [1] 2249
# > sum(hochgerner5k_2018_counts[9198,]>1)
# [1] 22
# > sum(hochgerner5k_2018_counts[2249,]>1)
# [1] 1054
# > sum((hochgerner5k_2018_counts[2249,]>1) & (hochgerner5k_2018_counts[9198,]>1) )
# [1] 1
```

#visualize PC3 on the feature plot


If you look at the commented code above you can see that only 1 cell expresses both Penk and Cck so it is a very strong marker.


Cembroskiw and Mannon (2018) notes that within a single cell type discrete clusters may not always be the best way to think about distiguishing cells.  Continuous cell states such as those underlying neuroplasticity may be at play and confused for distinct cells states.  In this cases it may be useful to just look at the compoenents of PCA and see which ones are most loaded with Jeager DEGs.  As 200 out of 900 of the Jeager DEGs were in the genes included by seurat one or two of these PCs may be significant sources of variation.  Seperating "static" or slower transcriptomic processes from faster acting IEG activity is a big challenge for this project and may be beyond the scope of a masters thesis.  But basically if certain cell types are
more responsive than others to certian stimulations their particular response will get picked up, but seperating this from 
the time course of their cell type specific activity may be challenging.  Without good time course data it is simply not
possible to make claims about these processes.  There are certian computational methods that proport to be able to detect
these processes and some reasources may be worth following up on have been provided here.


In particular it may be worht stuyding manifold learning methods.  

Moon, K. R., Stanley III, J. S., Burkhardt, D., van Dijk, D., Wolf, G., & Krishnaswamy, S. (2018). Manifold learning-based methods for analyzing single-cell RNA-sequencing data. Current Opinion in Systems Biology, 7, 36-46.
https://www.sciencedirect.com/science/article/pii/S2452310017301877
-very relevant

Moon et al. (2018) cites the following paper seems relevant to what you want to know:
Nadler, B., Lafon, S., Coifman, R. R., & Kevrekidis, I. G. (2006). Diffusion maps, spectral 
clustering and reaction coordinates of dynamical systems. Applied and Computational Harmonic Analysis, 21(1), 113-127.

and this cites a ton of other authors on this subject:  
3] D. Givon, R. Kupferman, A. Stuart, Extracting macroscopic dynamics: Model problems and algorithms, Nonlinearity 17 (2004) R55–R127.
[4] W. Huisinga, C. Best, R. Roitzsch, Ch. Schütte, F. Cordes, From simulation data to conformational ensembles: Structure and dynamics based
methods, J. Comput. Chem. 20 (1999) 1760–1774.
[5] W. Huisinga, Ch. Schütte, A.M. Stuart, Extracting macroscopic stochastic dynamics: Model problems, Comm. Pure Appl. Math. 56 (2003)
234–269.

This also gives a R package you can use: 
Zhao, J., Jaffe, A., Li, H., Lindenbaum, O., Sefik, E., Jackson, R., ... & Kluger, Y. (2021). Detection of differentially abundant cell subpopulations in scRNA-seq data. Proceedings of the National Academy of Sciences, 118(22).
https://www.pnas.org/doi/10.1073/pnas.2100293118
Rpackage: https://github.com/KlugerLab/DAseq
-does no limit itself to clusters
-can idenity differential abundance as well

Another method that looks good:
Skinnider, Michael A., Jordan W. Squair, Claudia Kathe, Mark A. Anderson, Matthieu Gautier, Kaya JE Matson, Marco Milano et al. "Cell type prioritization in single-cell data." Nature biotechnology 39, no. 1 (2021): 30-34.

More data to pick through:
Zhang, T. Y., Keown, C. L., Wen, X., Li, J., Vousden, D. A., Anacker, C., ... & Meaney, M. J. (2018).
Environmental enrichment increases transcriptional and epigenetic differentiation
between mouse dorsal and ventral dentate gyrus. Nature communications, 9(1), 1-11.
-Could look for markers of interest?
-Could this explain some of our results?



Buettner, F., Natarajan, K. N., Casale, F. P., Proserpio, V., Scialdone, A., Theis, F. J., ... & Stegle, O. (2015). Computational analysis of cell-to-cell heterogeneity in single-cell RNA-sequencing data reveals hidden subpopulations of cells. Nature biotechnology, 33(2), 155-160.

https://www.nature.com/articles/nbt.3102?report=reader

Kharchenko, P. V. (2021). The triumphs and limitations of computational methods for scRNA-seq. Nature Methods, 18(7), 723-732.
https://www.nature.com/articles/s41592-021-01171-x

In particular this tool CellChat may be interesting, it would be espescially interesting to use on the
Chen data as there may be connectivity data available to benchmark its observations

https://www.nature.com/articles/s41467-021-21246-9
Jin, S., Guerrero-Juarez, C. F., Zhang, L., Chang, I., Ramos, R., Kuan, C. H., ... & Nie, Q. (2021). Inference and analysis of cell-cell communication using CellChat. Nature communications, 12(1), 1-20.

https://github.com/sqjin/CellChat

Cressie, N., & Wikle, C. K. (2015). Statistics for spatio-temporal data. John Wiley & Sons.

Wikle, C. K. (2015). Modern perspectives on statistics for spatio‐temporal data. Wiley Interdisciplinary Reviews: Computational Statistics, 7(1), 86-98.


Please read the documentation of the dimensitonality reduction slots in Seurat objects:
https://satijalab.org/seurat/archive/v3.0/dim_reduction_vignette.html
-this page also described how to add custom dimensionality reductions to the seurat object which is something I may want to do as well
run ICA:
https://satijalab.org/seurat/reference/runica

```{r}
Hoch5k.GCadult <- RunICA(Hoch5k.GCadult, nics = 7)
```
```{r}
DimPlot(Hoch5k.GCadult, reduction = "ica")
```
```{r}
FeaturePlot(Hoch5k.GCadult, reduction = "ica", features = c("Penk", "Cck"), blend = TRUE)
```

```{r}
FeaturePlot(Hoch5k.GCadult, reduction = "ica", features = c("Penk", "Malat1"), blend = TRUE)
```


```{r}
DimHeatmap(Hoch5k.GCadult, reduction = "ica", dims = 1:2)

```

tSNE and UMAP are no use for this as they do not return feature loadings.  They have little to say about the underlying structure and are purely good for visualization.  Though Here I would say that PCA is even doing a better job and ICA possibly even better.



## Getting Loadings out of Components


The plan here is to use some reduction, be it ICA, PCA or a Laplacian Embedding and then use calculate the absolute loading
of Jeager DEGs, possibly by time course as well and then use the non Jeager DEGs as markers in the analysis focused on jsut the DEGs known to be in the process in question.  This is a methods to obeserve interactions between fast time scale transcriptional processes and slow process transcriptional changes.  We could for instance tyr to use it to see how Memory Trace and Circadian Rhythms interact (Costa, RObinson-Rechavi, & Rupperger, 2020) as we have a dataset that could allow us to examine two "fast scale" processes circadian rythms and memory trace over the course of a developmental timespan with slower dynamics.  Diffusion mapping, i.e. pseudo-time methods would be ideal for this (Hagverdi Buettner, & Theis, 2015; Coifman et al., 2005).  Combining this approach with a more supervised approach could be done using autoencoders (Tran et al., 2021).

Costa, S. S. F., Robinson-Rechavi, M., & Ripperger, J. A. (2020). Single-cell transcriptomics allows novel insights into aging and circadian processes. Briefings in Functional Genomics, 19(5-6), 343.
https://www.ncbi.nlm.nih.gov/labs/pmc/articles/PMC7716582/ 


Coifman, R. R., Lafon, S., Lee, A. B., Maggioni, M., Nadler, B., Warner, F., & Zucker, S. W. (2005). Geometric diffusions as a tool for harmonic analysis and structure definition of data: Diffusion maps. Proceedings of the national academy of sciences, 102(21), 7426-7431.

Haghverdi, L., Buettner, F., & Theis, F. J. (2015). Diffusion maps for high-dimensional single-cell analysis of differentiation data. Bioinformatics, 31(18), 2989-2998.

Tran, D., Nguyen, H., Tran, B., La Vecchia, C., Luu, H. N., & Nguyen, T. (2021). Fast and precise single-cell data analysis using a hierarchical autoencoder. Nature communications, 12(1), 1-10.

```{r}
#script to get and rank gene's in PCs
testPC <- Hoch5k.GCadult[['pca']][,1]
testPC <-data.frame(testPC)
testPC$gene <- rownames(testPC)
testPC$abs <- abs(testPC$PC_1)
testPC$sign <- testPC$PC_1 > 0

N = 20
topN <- testPC %>%
  slice_max(n = N, order_by = abs) %>%
  group_by(sign)

topN
    
```

Alternatively Seurat does posses means to detect cell state, should read the following publication: 
Hao, Y., Hao, S., Andersen-Nissen, E., Mauck III, W. M., Zheng, S., Butler, A., ... & Satija, R. (2021). Integrated analysis of multimodal single-cell data. Cell, 184(13), 3573-3587.

It may simply be the case that transcriptomically identifying these states will always be a challenge

And see if it may hold some answers.



```{r}
print(Hoch5k.GCadult[["pca"]], dims = 1:7, nfeatures = 50)
```
Lamia, K. A., Papp, S. J., Yu, R. T., Barish, G. D., Uhlenhaut, N. H., Jonker, J. W., ... & Evans, R. M. (2011). Cryptochromes mediate rhythmic repression of the glucocorticoid receptor. Nature, 480(7378), 552-556.

<br>
<br>
It may make sense to remove these JeagerDEG genes from the list entirely and try to see how it affects the clustering and cluster just using the DEGs.  Most Jeager DEGs were upregulated but some were downregulated so it may be useful to check their loading sign as well.  I cannot guarantee which loading Penk and Rgs4 will show up in but it will always be in the top 3 PCs.
<br>
<br>
```{r}
VizDimLoadings(Hoch5k.GCadult, dims = 1, reduction = "pca")
```
```{r}
VizDimLoadings(Hoch5k.GCadult, dims = 2, reduction = "pca")
```

```{r}
VizDimLoadings(Hoch5k.GCadult, dims = 3, reduction = "pca")
```



* get the cembrowski and mannon  citation its the one on continuous variation within cell types


So overall while the Jeager DEGs that are included in the analysis are present in many of the most important PCs they are not clearly marking some cells as engram cells with the rest of the gene's present.  We could consider withholding the genes that
are Jeager DEGs from the initial clustering as well and see how they interact, or alternatively perform CCA on the two data 
sets then cluster over two processes and get cell state markers and cell subtype markers.  Alternatively look into use the graph laplacian to look for "reaction coordinates" (Moon et al., 2018; Nadler et al., 2006).  There are also non-linear generalizations of CCA we could try as well.




Clustering just within the Jeager DEGs
---


Only 200 of the ~900 JeagerDEGs were included in the most variable gene's and seurats filtering of features, while good
for large data sets to identify specific processes may be obscuring gene's we know to be important in cell state transitions.
As for the clustering in the data above I would like to take a note from Cembrowski & Mannon et al. (2020)  about cell state transitions being more subtle than clustering and that prehaps clustering into discrete classes is not a good way to understand a more homogenous population.  We can alos try including cells from across different ages but we will need to control for age effects.


```{r}
Hoch5k.GCadult.onlyJeagerDEGs <- CreateSeuratObject(
  counts = hochgerner5k_2018_counts[Jeager.DEGs$hoch5k.idx, hoch5k.GC_Adult.p35.idx],
  meta.data = hochgerner5k_2018_meta[ hoch5k.GC_Adult.p35.idx, ])

Hoch5k.GCadult.onlyJeagerDEGs
```



```{r}
Hoch5k.GCadult.onlyJeagerDEGs <- NormalizeData(Hoch5k.GCadult.onlyJeagerDEGs) # Seurat normalization has many steps see publication
# https://www.frontiersin.org/articles/10.3389/fgene.2020.00041/full describes and compares it to others
Hoch5k.GCadult.onlyJeagerDEGs <- FindVariableFeatures(Hoch5k.GCadult.onlyJeagerDEGs)
Hoch5k.GCadult.onlyJeagerDEGs
```
So now when we exclude the non Jeager IEGs we get more signal from our desired gene's.


```{r}
top10 <-  head(VariableFeatures(Hoch5k.GCadult.onlyJeagerDEGs), 10)
plot1 <- VariableFeaturePlot(Hoch5k.GCadult.onlyJeagerDEGs)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2 + ggtitle("Variable") +  theme(plot.title = element_text(hjust = 0.5))
```

```{r}
top10 <-  head(VariableFeatures(Hoch5k.GCadult.onlyJeagerDEGs), 10)
```




```{r}
all.genes <- rownames(Hoch5k.GCadult.onlyJeagerDEGs)
Hoch5k.GCadult.onlyJeagerDEGs <- ScaleData(Hoch5k.GCadult.onlyJeagerDEGs, features = all.genes) 

Hoch5k.GCadult.onlyJeagerDEGs <- RunPCA(Hoch5k.GCadult.onlyJeagerDEGs, 
                                        features = all.genes,
                                        npcs = 20) 
#note that excluding genes using variable features meant that inhba did not show up in the markers, to me that was unacceptbale
```
So here we can see several gradients of IEGs and engram associated genes from across the time course data.  Whether or not these reflect cell type specific processes or a particular point in a time course of one cell type, or a mix of these possibilities is difficult to assess.  Possibly include a figure showing these three possibilities, time courses as color gradients, cell types as shapes and behavioral exposures as letters.


```{r}
#Dsicovering dimensionality of these activity markers
Hoch5k.GCadult.onlyJeagerDEGs <- JackStraw(Hoch5k.GCadult.onlyJeagerDEGs, num.replicate = 100)
Hoch5k.GCadult.onlyJeagerDEGs <- ScoreJackStraw(Hoch5k.GCadult.onlyJeagerDEGs, dims = 1:20)

```


```{r}
JackStrawPlot(Hoch5k.GCadult.onlyJeagerDEGs, dims = 1:20)
```



```{r}

ElbowPlot(Hoch5k.GCadult.onlyJeagerDEGs) +
  ggtitle("PCs of Activity DEGs Only") +
  theme(plot.title = element_text(hjust = 0.5))

```


```{r}
#Clustering, we will visualize in UMP later
Hoch5k.GCadult.onlyJeagerDEGs <- FindNeighbors(Hoch5k.GCadult.onlyJeagerDEGs, dims = 1:5)
Hoch5k.GCadult.onlyJeagerDEGs <- FindClusters(Hoch5k.GCadult.onlyJeagerDEGs, resolution = 0.2)

```

```{r}
onlyJeagerDEGs.markers <- FindAllMarkers(Hoch5k.GCadult.onlyJeagerDEGs, min.pct = 0.25, logfc.threshold = 0.25)
```

```{r}
p <- DimPlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'pca') 

p + ggtitle("Two Markerless Clusters") + theme(plot.title = element_text(hjust = 0.5))
```



```{r}
top2markers <- onlyJeagerDEGs.markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
top2markers
```



```{r}
top5markers <- onlyJeagerDEGs.markers %>%
    group_by(cluster) %>%
    slice_max(n = 5, order_by = avg_log2FC)

top5markers 
```
```{r}
#Engram cell heatmap
DoHeatmap(Hoch5k.GCadult.onlyJeagerDEGs, features = top5markers$gene)
```
```{r}
table(Hoch5k.GCadult.onlyJeagerDEGs@meta.data$seurat_clusters)
```
So what we see above is somewhat arbitrary but I have good experimental reasons to believe these are engram cells.
They are a rare cell type expressing known IEG markers that are strongly active following activity.  Although i feel the main issue that arrises is that we are trying to cluster them.  The gradients (components) of this analysis itself may be informative.

```{r}
p <- DimPlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'pca')
p + ggtitle("Putative Engram Cells") + theme(plot.title = element_text(hjust = 0.5))

table()
```
```{r}
# this does not work so well looks ugly
Inhba.p <- FeaturePlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'pca', feature = "Inhba")
Rgs4.p <- FeaturePlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'pca', feature = "Rgs4")
Nptx2.p <- FeaturePlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'pca', feature = "Nptx2")

Inhba.p + Rgs4.p + Nptx2.p
```
```{r}
FeaturePlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'pca', feature = c("Inhba", "Rgs4") )
```
```{r}
FeaturePlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'pca', feature = c("Nptx2", "Lingo1") )
```



### Checking Penk Expression

```{r}
# renaming clsuters and checking for penk expression in the putative engram cells
engram2PC.clusters <- Hoch5k.GCadult.onlyJeagerDEGs@meta.data$seurat_clusters
new.cluster.ids <- c("Inactive_1", "Inactive_2","Inactive_3", "Inactive_4,", "Putative_EngramCells")
names(new.cluster.ids) <- levels(Hoch5k.GCadult.onlyJeagerDEGs)
names(engram2PC.clusters) <- levels(Hoch5k.GCadult.onlyJeagerDEGs)
Hoch5k.GCadult.onlyJeagerDEGs <- RenameIdents(Hoch5k.GCadult.onlyJeagerDEGs, new.cluster.ids)
DimPlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = "pca", label = TRUE, pt.size = 0.5) + NoLegend()

Hoch5k.GCadult@meta.data$onlyJeagerDEGlabel <- Hoch5k.GCadult.onlyJeagerDEGs@meta.data$seurat_clusters
#Hoch5k.GCadult@meta.data$
```


```{r}
# > which(rownames(Hoch5k.GCadult)=="Penk")
# [1] 9198
# > which(rownames(Hoch5k.GCadult)=="Cck")
# [1] 2249
# > which(rownames(Hoch5k.GCadult)=="Malat1")
# [1] 7375
Cck.loc = 2249


Penk.loc = 9198

test = hochgerner5k_2018_counts[, hoch5k.GC_Adult.p35.idx]
sum(test[9198, which(Hoch5k.GCadult@meta.data$onlyJeagerDEGlabel == "4")])
                                
x = as.numeric(test[Penk.loc , which(Hoch5k.GCadult@meta.data$onlyJeagerDEGlabel == "4")]) 
y = as.numeric(test[Penk.loc , which(Hoch5k.GCadult@meta.data$onlyJeagerDEGlabel != "4")])

wilcox.test(x, y)

nonegramcelltotal = sum(Hoch5k.GCadult@meta.data$onlyJeagerDEGlabel != "4")

penk_engramcells = sum(x>1)
penk_nonengramcells = sum(y>1)
#now with Malat1 
Malat1.loc = 7375

x = as.numeric(test[Malat1.loc, which(Hoch5k.GCadult@meta.data$onlyJeagerDEGlabel == "4")]) 
y = as.numeric(test[Malat1.loc, which(Hoch5k.GCadult@meta.data$onlyJeagerDEGlabel != "4")])

wilcox.test(x, y)


#How many Penk expressing cells are also in the 3rd quantile of Malat1 expression?
sum((test[Penk.loc,]>1)&(test[Malat1.loc,]>220))

#How many Penk expressing cells in the 3rd quantile of Malat expression are Engram cells?
sum(which((test[Penk.loc,]>1)&(test[Malat1.loc,]>220)) %in% which(Hoch5k.GCadult@meta.data$onlyJeagerDEGlabel == "4"))
#Answer 6/8 which means 6/7 penk expressing Engram cells are also in the 3rd quantile of Malat1 expression


#Malat expressing cells in the 3rd quantile are over represented in the engram cells
# > summary(y)
#    Min. 1st Qu.  Median    Mean 3rd Qu.    Max. 
#    39.0   110.2   141.0   143.6   170.8   414.0 

nonegramcelltotal = sum(Hoch5k.GCadult@meta.data$onlyJeagerDEGlabel != "4")
penk_engramcells = sum(x>1)

# > sum(x>171) # engram cells in top 3rd quantile of Malat expression
# [1] 70

dim(hochgerner5k_2018_counts)


```

```{r}
# Plotting for wilcoxin
df = data.frame(t(test[c(Malat1.loc,Penk.loc,Cck.loc),]))
df$Penk <- as.numeric(df$Penk)
df$Malat1 <- as.numeric(df$Malat1)
df$Cck <- as.numeric(df$Cck)
df$EngramStatus <- as.character(Hoch5k.GCadult@meta.data$onlyJeagerDEGlabel)
df$EngramStatus = as.factor(df$EngramStatus == "4")
levels(df$EngramStatus) <- c("InactiveCell","EngramCell")
df$PenkStatus <- as.factor(df$Penk>1)
levels(df$PenkStatus) <- c("Penk-", "Penk+")

# plotting this looks ugly jsut stick to reporting the data verbally
p <- ggboxplot(df, x = "EngramStatus", y = "Cck",
          color = "EngramStatus", palette = "jco",
          add = "jitter") 
#  Add p-value
p <- p + stat_compare_means()
# Change method
p + stat_compare_means(method = "t.test")


ggpaired(df, x = "EngramStatus", y = "Penk",
         color = "EngramStatus", line.color = "gray", line.size = 0,
         palette = "jco")+
  stat_compare_means(paired = TRUE)
```






```{r}
##USEFUL CODE  BUT NOT FOR PLOTTING
Hoch5k.GCadult@meta.data$engramlabel <- engram2PC.clusters

engramlabel.withdata <- data(Hoch5k.GCadult)

FeaturePlot(Hoch5k.GCadult, reduction = 'pca', features = )


features <- c("Penk", "Malat1", "Cck", "Inhba", "Rgs4", "Fos")

Hoch5k.GCadult.test@meta.data$threePC_clusters <- seurat_clusters.3dims
Hoch5k.GCadult.test@meta.data$sixPC_clusters <- seurat_clusters.6dims

Idents(Hoch5k.GCadult.test) <- Hoch5k.GCadult.test@meta.data$threePC_clusters
RidgePlot(Hoch5k.GCadult.test, features = features, ncol = 2)
VlnPlot(Hoch5k.GCadult.test, features = features)

```



```{r}

which(rownames(Hoch5k.GCadult)=="Penk")
which(rownames(Hoch5k.GCadult)=="Cck")
which(rownames(Hoch5k.GCadult)=="Malat1")
```

which(rownames(Hoch5k))



### Trying it with some markers from the other data

Nothing useful really shows up malat and Cck show up but penk does not

```{r}
#onlyJeagerDEGs but we will include Penk, Malat1 and Cck in the gene list

#qwert to find this section

include.me <- c(Jeager.DEGs$hoch5k.idx, 9198, 2249, 7375)

DEGs.and.mrks <- CreateSeuratObject(
  counts = hochgerner5k_2018_counts[include.me, hoch5k.GC_Adult.p35.idx],
  meta.data = hochgerner5k_2018_meta[ hoch5k.GC_Adult.p35.idx, ])

all.genes <- rownames(DEGs.and.mrks)
DEGs.and.mrks <- ScaleData(DEGs.and.mrks, features = all.genes) 

DEGs.and.mrks <- FindVariableFeatures(DEGs.and.mrks)

DEGs.and.mrks <- RunPCA(DEGs.and.mrks,
                        features = all.genes,
                        npcs = 20)
```

```{r}
top10 <-  head(VariableFeatures(DEGs.and.mrks), 10)
plot1 <- VariableFeaturePlot(DEGs.and.mrks)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```




```{r}
DEGs.and.mrks <- JackStraw(DEGs.and.mrks, num.replicate = 100)
DEGs.and.mrks <- ScoreJackStraw(DEGs.and.mrks, dims = 1:20)

ElbowPlot(DEGs.and.mrks)
```


```{r}
DEGs.and.mrks <- FindNeighbors(DEGs.and.mrks, dims = 1:3)
DEGs.and.mrks <- FindClusters(DEGs.and.mrks, resolution = 0.2)
```


```{r}
gradient.markers <- FindAllMarkers(DEGs.and.mrks, min.pct = 0.25, logfc.threshold = 0.25)

top10markers <- gradient.markers  %>%
    group_by(cluster) %>%
    slice_max(n = 10, order_by = avg_log2FC) %>%


gradient.markers %>%
  filter(p_val_adj<0.05) %>%
    group_by(cluster) %>%
    slice_max(n = 5, order_by = avg_log2FC)
```

```{r}

for(i in c(0:3)){
  mrks.temp <- gradient.markers$gene[which(gradient.markers$cluster==as.character(i))]
  mrks.temp[which(mrks.temp %in% Jeager.DEGs$GeneID)]
  Jeager.DEGs$group.number[Jeager.DEGs$GeneID %in% mrks.temp]
  Jeager.DEGs$group[Jeager.DEGs$GeneID %in% mrks.temp]
}

```





```{r}
table(DEGs.and.mrks)
```



```{r}
DEGs.and.mrks <- RunICA(DEGs.and.mrks)
```





Contextualizing the results
---







```{r}
PC_Loadings.onlyJeagerDEGs <- data.frame(Loadings(Hoch5k.GCadult.onlyJeagerDEGs, reduction = "pca")[, 1:4])

rank_property_in_loading<- function(dataframe.of.loadings){
  
  
}


#script to get and rank gene's in PCs
testPC <- Hoch5k.GCadult.onlyJeagerDEGs[['pca']][,1]
testPC <-data.frame(testPC)
testPC$gene <- rownames(testPC)
testPC$abs <- abs(testPC$PC_1)
testPC$sign <- testPC$PC_1 > 0

N = 20
topN <- testPC %>%
  slice_max(n = N, order_by = abs) %>%
  group_by(sign)

topN

#Jeager.DEGs$hoch5k.idx <- which(rownames(hochgerner5k_2018_counts) %in% Jeager.DEGs$GeneID)

#thought, when you look at the absolute values and then cut off the top it may be possible to get and idea
# of how "polarized a gradient is, i.e. how many of its laodings are caught up on one side of it,"
# most of the PC1 genes in the onlyJeagerDEGs data are all positive

data.frame(Loadings(Hoch5k.GCadult.onlyJeagerDEGs, reduction = "pca")[, 1:4])  
  
  

```






```{r}
Hoch5k.GCadult.onlyJeagerDEGs <- RunUMAP(Hoch5k.GCadult.onlyJeagerDEGs, dims = 1:4)

DimPlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction ="umap")
```


```{r}
DimPlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction ="pca")
```


### Trying out ICA for visualization 


```{r}
Hoch5k.GCadult.onlyJeagerDEGs <- RunICA(Hoch5k.GCadult.onlyJeagerDEGs)
```
In IC_1  NOTE THAT THIS IS RUN SPECIFICALLY ON THE JEAGER ET AL. GENEs 
Chang, S., Bok, P., Tsai, C. Y., Sun, C. P., Liu, H., Deussing, J. M., & Huang, G. J. (2018). NPTX2 is a key component in the regulation of anxiety. Neuropsychopharmacology, 43(9), 1943-1953.
-Npt2x supresses anxiety

Pinzari, C. A., Kang, L., Michalak, P., Jermiin, L. S., Price, D. K., & Bonaccorso, F. J. (2020). Analysis of genomic sequence data reveals the origin and evolutionary separation of Hawaiian hoary bat populations. Genome biology and evolution, 12(9), 1504-1514.
BAZ1A is mutated in bat populations showing differences in foragin behaviours

Conith, M. R., Hu, Y., Conith, A. J., Maginnis, M. A., Webb, J. F., & Albertson, R. C. (2018). Genetic and developmental origins of a unique foraging adaptation in a Lake Malawi cichlid genus. Proceedings of the National Academy of Sciences, 115(27), 7063-7068.
-smad4 is mutated in a fish species with different foragin behaviours, it was implicated with snout shape in this study which
is useful for certain behaviours 

Wang, C. N., Gong, S. N., Guan, W., Wang, J. L., Gao, T. T., Wang, Y., ... & Jiang, B. (2020). Hippocampal overexpression of chordin protects against the chronic social defeat stress-induced depressive-like effects in mice. Brain research bulletin, 158, 31-39.
-smad4 suppresses stress

Faguet, J. (2014). The role of neuropeptide signaling in C. elegans food sensation and social behavior. University of California, San Francisco.
-in celegans smad is assocaited with foraging, in food rich environments 

Fear memories are not replayed during theta osscilations:  Feng, H., Su, J., Fang, W., Chen, X., & He, J. (2021). The entorhinal cortex modulates trace fear memory formation and neuroplasticity in the lateral amygdala via cholecystokinin. bioRxiv.
-Cck is expressed in fear memories in entorhinal cortex in neurons that project to Cck expressing neurons in amygdala,
this circuit is likely driving fear memory formation 

Feng, H., Su, J., Fang, W., Chen, X., & He, J. (2021). The entorhinal cortex modulates trace fear memory formation and neuroplasticity in the mouse lateral amygdala via cholecystokinin. Elife, 10, e69333.  

Smagin, D. A., Kovalenko, I. L., Galyamina, A. G., Bragin, A. O., Orlov, Y. L., & Kudryavtseva, N. N. (2016). Dysfunction in ribosomal gene expression in the hypothalamus and hippocampus following chronic social defeat stress in male mice as revealed by RNA-Seq. Neural Plasticity, 2016.
Rpl7 is implicated in chronic social stress in rodents
-RNA seq study

Sagehashi, N., Obara, Y., Maruyama, O., Nakagawa, T., Hosoi, T., & Ishii, K. (2022). Insulin enhances gene expression of Midnolin, a novel genetic risk factor for Parkinson's disease, via ERK, PI3-kinase and multiple transcription factors in SH-SY5Y cells. Journal of Pharmacology and Experimental Therapeutics.
-Midn is expressede in response to insulin 

Pitychoutis, P. M., Sanoudou, D., Papandreou, M., Nasias, D., Kouskou, M., Tomlinson, C. R., ... & Papadopoulou-Daifoti, Z. (2014). Forced swim test induces divergent global transcriptomic alterations in the hippocampus of high versus low novelty-seeker rats. Human genomics, 8(1), 1-13.
-smad7 is differentially regulated between two novelty seeking rodent lines who differ in amount of observed locomotor activity


Barbier, M., Chometton, S., Pautrat, A., Miguet-Alfonsi, C., Datiche, F., Gascuel, J., ... & Risold, P. Y. (2020). A basal ganglia-like cortical–amygdalar–hypothalamic network mediates feeding behavior. Proceedings of the National Academy of Sciences, 117(27), 15967-15976.
-a basal ganglia like feeding network exists between cortex, amygdala and hypothalamus and may have split from the basal ganglia

Flati, T., Gioiosa, S., Chillemi, G., Mele, A., Oliverio, A., Mannironi, C., ... & Castrignanò, T. (2020). A gene expression atlas for different kinds of stress in the mouse brain. Scientific Data, 7(1), 1-18.
-visual cortex, dentate gyrus and other regions have elevated Cck expression in stress conditions

https://www.sciencedirect.com/science/article/pii/S0079612307630155?casa_token=UAsB4LZLpDoAAAAA:lprJ_G_Hctq5HU6cj8dzHG_ridQX-cc9w6wB1tJI7hzKPxKHm0685_JZ-Eh-Ftk8XztkUO-EaYo


Shoham, S., Marcus, E. L., Avraham, Y., & Berry, E. M. (2000). Diet restriction increases enkephalin-and dynorphin-like immunoreactivity in rat brain and attenuates long-term retention of passive avoidance. Nutritional neuroscience, 3(1), 41-55.
-blood sugar is strongly negatively correlated with Penk expression and low engergy diets induce changes in stress related memory tasks but not food rewarded tasks


https://www.nature.com/articles/s41586-021-03811-w


```{r}
DimPlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'ica')
```
```{r}
FeaturePlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'ica', 
            features = c("Inhba", "Hs3st1"), blend = TRUE)
```




```{r}
FeaturePlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'pca', 
            features = c( "Inhba", "Rpl4"), blend = TRUE)

```

```{r}
FeaturePlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction = 'pca', 
            features = c( "Rgs4", "Inhba"), blend = TRUE)

```


I am satisfied that we have identified a cluster of engram cells in cluster 3.  I am concerned this method is not
robust and reproducible, it took a fair bit of fiddling with it maybe for that reason a classifier will be better.  Though
I am not entirely willing to just give up on an unsupervised method for doing this.  May be with the right embedding and better data one could find a better method.

```{r}
Hoch5k.GCadult@meta.data$JeagerDEGs_clusters_3PCs <- Hoch5k.GCadult.onlyJeagerDEGs@meta.data$seurat_clusters
```


Definately worth talking about this with regard to the DEG expression across hippocampus: https://pubmed.ncbi.nlm.nih.gov/29875406/

Shows lack of stability in CA1 and Ca2 representation in calcium imaging data.



```{r}
onlyJeagerDEGs.markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
```



```{r}
p <- DimPlot(Hoch5k.GCadult.onlyJeagerDEGs, reduction ="umap")

p <- p + ggtitle("Adult DGCs clustered with Engram DEGs (UMAP)") +
  theme(plot.title = element_text(hjust = 0.5)) +
  xlab("Dim 1") +
  ylab("Dim 2")
  
  
p
# FOR REFERENCE: 
# ggplot(df, aes(x=fos, color = cond)) + 
#   geom_histogram(fill =  "white") +
#   xlab("Fos Reads") +
#   ylab("Frequency") +
#   theme_classic() + 
#   theme(legend.title = element_blank()) +
#   ggtitle("Freqency of Fos Counts in Jeager Groups")+
#   theme(plot.title = element_text(hjust = 0.5))+
#   theme(axis.text = element_text(face = "bold"))
  
```


<br>
<br>
So what we get out of this is some kind of graded expression within the population.  It would be interesting to try to see which cells have more of which gene group in them.  This is similar to what was described by Dr. Cembrowski's group in Erwin et al., (2020) see figure 5 A in particular.  I tried in with 3 dims and I think it looks better with 4 more graded less like swiss cheese.  it would be interesting to see how Jeager's group labels map onto this.
<br>
<br>

<br>
<br>
The other clusters are more or less intermixed while the *Inhba* labelled neurons are clearly moving along PC1.
<br>
<br>

```{r}
JeagerDEG.markers <- FindAllMarkers(Hoch5k.GCadult.onlyJeagerDEGs, only.pos = FALSE, min.pct = 0.25, logfc.threshold = 0.25)
JeagerDEG.markers$JeagerGrouping <- Jeager.DEGs$group[Jeager.DEGs$GeneID %in% JeagerDEG.markers$gene]

JeagerDEG.markers %>%
    group_by(cluster) %>%
    slice_max(n = 5, order_by = avg_log2FC)
```
```{r}
JeagerDEG.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10



DoHeatmap(Hoch5k.GCadult.onlyJeagerDEGs, features = c(top10$gene, "Inhba", "Fos"), label = FALSE) + 
  theme(axis.text.y = element_text(face = "bold")) +
  theme(legend.title = element_text(face = "bold")) + 
  ggtitle("Hochgerner P35 Mature DCGs Clustered with Jeager Engram DEGs")

```
<br>
<br>


NEXT STEPS
---
<br>
<br>
Cells from Hochgerner et al., (2018) have all passed theri quality controls, I may need to include DGC markers as a sort of control to see if they drive any clustering differently or if Jeager markers are over represented in a particular cell type.
Should also perform this analysis on immature granule cells as well and see if the same markers show up or slightly different ones.rm
<br>
<br>
View this tutorial later:  https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionIV/lessons/SC_marker_identification.html

May have good information for labeling FeaturePlot function in seurat with gene expression information.

```{r}
#saving our object and it's data, commented out so its doesn't run everytime

# write.csv(Hoch5k.GCadult.markers, "SeuratAnalysesData/HochgernerGDC_Adult_marker.csv")
# 
# remotes::install_github("mojaveazure/seurat-disk")
# library(SeuratDisk)
# SaveH5Seurat(Hoch5k.GCadult, filename = "SeuratAnalysesData/HochgernerDGC_Adult_P35.h5Seurat", overwrite = TRUE)
```

<br>
<br>

References
---
<br>
<br>
Erwin, S. R., Sun, W., Copeland, M., Lindo, S., Spruston, N., & Cembrowski, M. S. (2020). A sparse, spatially biased subtype of mature granule cell dominates recruitment in hippocampal-associated behaviors. Cell Reports, 31(4), 107551. 
<br>
<br>
Giorgi, C., & Marinelli, S. (2021). Roles and Transcriptional Responses of Inhibitory Neurons in Learning and Memory. Frontiers in Molecular Neuroscience, 113.
<br>
<br>
Jaeger, B. N., Linker, S. B., Parylak, S. L., Barron, J. J., Gallina, I. S., Saavedra, C. D., ... & Gage, F. H. (2018). A novel environment-evoked transcriptional signature predicts reactivity in single dentate granule neurons. Nature communications, 9(1), 1-15.
<br>
<br>
Joglekar, A., Prjibelski, A., Mahfouz, A., Collier, P., Lin, S., Schlusche, A. K., ... & Tilgner, H. U. (2021). A spatially resolved brain region-and cell type-specific isoform atlas of the postnatal mouse brain. Nature Communications, 12(1), 1-16.
<br>
<br>
Habib, N., Li, Y., Heidenreich, M., Swiech, L., Avraham-Davidi, I., Trombetta, J. J., ... & Regev, A. (2016). Div-Seq: Single-nucleus RNA-Seq reveals dynamics of rare adult newborn neurons. Science, 353(6302), 925-928.
<br>
<br>
Hochgerner, H., Zeisel, A., Lönnerberg, P., & Linnarsson, S. (2018). Conserved properties of dentate gyrus neurogenesis across postnatal development revealed by single-cell RNA sequencing. Nature neuroscience, 21(2), 290-299.
<br>
<br>
Klosen, P., Lapmanee, S., Schuster, C., Guardiola, B., Hicks, D., Pevet, P., & Felder‐Schmittbuhl, M. P. (2019). MT1 and MT2 melatonin receptors are expressed in nonoverlapping neuronal populations. Journal of pineal research, 67(1), e12575.
<br>
<br>
Raivich, G., Bohatschek, M., Da Costa, C., Iwata, O., Galiano, M., Hristova, M., ... & Behrens, A. (2004). The AP-1 transcription factor c-Jun is required for efficient axonal regeneration. Neuron, 43(1), 57-67.
<br>
<br>


