---
title: "Jeager Engram Cell Gene signatures in other datasets"
author: "Angus Campbell"
date: "11/02/2022"
output: github_document
---

```{r include = FALSE}
#get libraries
setwd("C:/Users/angus/Desktop/PavLabEngrams/EngramCellClassifier")

library(tidyverse)
library(GEOquery)
library(AnnotationDbi)
library(randomForest)
library(data.table)
library(reshape2)
library(FactoMineR)
library(factoextra)
library(Rtsne)
library(dplyr)
library(Seurat) #installed with remotes
library(stringr)
library(patchwork)
library(metap)
library(cqn)
library(GeoTcgaData)
library(ggplot2)
library(tictoc)
library(pheatmap)
library(SeuratDisk) # installed with remotes
#reticulate::py_install(packages ='umap-learn')

#Seurat reference manual: https://cloud.r-project.org/web/packages/Seurat/Seurat.pdf
```

NOTES: Line 168 needs fixing, we need to filter cells out with all the gene's before restricting to just the Jeager DEGs,
then you need to change the stats as well, make sure they are being corrected by Benjamini-Hochberg


Introduction
---

Here I have used supplemental data 4 from Jeager et al., (2018) which is a list of the DEGs from Fos+ dentate granule cells (DGCs) at various time points vs Fos- DGCs at various time points.  They assigned these DEGs to qualitatively similar time courses.  While not rigorous this manual sissignment is still interesting to use, perhaps later we can redo this with a more rigorous exploration of the time course data.  Regardless I use this DEG list to cluster DGCs from the first experiment of Hochgerner et al., (2018).  This second dataset of ~5k cells was sampled from mlae and female mice across development.  The sample of interest for me were samples from post natal day 37 which is relatively close to the 7-8.75 week old mice (in days they are ~P49-P62) in Jeager.  Though not an ideal match it is better than the other dataset which contains 24k cells but range from embryonic mice to post natal day 120 and 132.  Jeager et al. (2018) was done entirely using female mice while Hochgerner was done with a mixture of males and females.

The Hochgerner dataset contains some peculiarities.  They have pooled cells between males and females to reduce batch effects, but its difficult to guess which cells came from which.  Hochgerner et al., (2018)  assigned sex to cells based on number of male vs female genes in each cell.  They provide metadata about the number fo males and females pooled for the batch each cell was drawn from but in the post natal days 24 and 35 in the Hoch5k data there are some that simply cannot be known.


### Jeager et al., (2018) meta data and counts

Load Jeager data.

```{r}
jeager2018_counts <- bind_cols(read.table('Jeager2018_GSE98679/GSE98679_count.txt.gz', header = TRUE, check.names = FALSE),
                               read.table('Jeager2018_GSE98679/GSE98679_v2_GSM3308862-GSM3309413_count.txt.gz', header = TRUE, check.names = FALSE))

jeager2018_meta <- read.csv('Jeager2018_GSE98679/SraRunTable.txt', header = TRUE)
jeager2018_meta = jeager2018_meta[c(1:46,599:912,47:598),] #we need to fix Jeager's files up a bit
rownames(jeager2018_meta) <- c(1:912)

jeager2018_meta$predicted_cell_type <- as.character(lapply(jeager2018_meta$predicted_cell_type, function(x) if (x=="") {"DG"} else {x}))
jeager2018_meta$predicted_cell_type <- lapply(jeager2018_meta$predicted_cell_type, function(x) if (x=="") {"DG"} else {x})

jeager2018_meta$fos_status <- as.factor(sapply(as.character(jeager2018_meta$source_name), function(y) if (grepl("_F_", y, fixed=TRUE)) "Fos+" else "Fos-"  ))
jeager2018_meta$fos_status[361:dim(jeager2018_meta)[1]] <- "Fos+"

jeager2018_meta$Mouse_Number[c(361:912)] <- jeager2018_meta$mousingle_number[c(361:912)]
jeager2018_meta <- jeager2018_meta %>% 
  dplyr::select(-mousingle_number)

#filtering out cells as per instructions in Jeager et al., 2018)
under4k <- sapply(jeager2018_counts, function(y) sum(length(which(y>0))))

filtered.idx <- as.numeric(which(under4k>4000))
filtered.idx <- order(c(filtered.idx,194))

jeager2018_counts <-jeager2018_counts[,filtered.idx]
jeager2018_meta <- jeager2018_meta[filtered.idx,]
```



### Getting a handle on realsitic Fos levels in Jeager

<br><br>
Quick check for HC DGCs with fos labelleing.
<br>
```{r}
which((jeager2018_meta$exposure=="HC")&(jeager2018_meta$fos_status=="Fos+"))
```
<br><br>
So no data on that but let's get an idea for what the IEG distribution was like in Jeager's experiments.
<br><br>
```{r}
jeager.fos.idx <- which(rownames(jeager2018_counts)=="Fos")
fosneg.HC <- which((jeager2018_meta$exposure=="HC")&(jeager2018_meta$fos_status=="Fos-"))

hist(as.numeric(jeager2018_counts[jeager.fos.idx, fosneg.HC]), 
     main = "Homecage Fos-AntiBd Negative nuclear Fos read distribution",
     xlab = "Fos Read Count")


```
<br>
```{r}
fosneg.1hr <- which((jeager2018_meta$exposure=="1hr")&(jeager2018_meta$fos_status=="Fos-"))


hist(as.numeric(jeager2018_counts[jeager.fos.idx, fosneg.1hr]), 
     main = "1 Hour pst NE Fos-AntiBd Negative nuclear Fos read distribution",
     xlab = "Fos Read Count")
```

```{r}
fospos.1hr <- which((jeager2018_meta$exposure=="1hr")&(jeager2018_meta$fos_status=="Fos+"))

hist(as.numeric(jeager2018_counts[jeager.fos.idx, fospos.1hr]),
     main = "1Hour post-NE Fos-AntiBd Positive nuclear Fos read distribution",
     xlab = "Fos Read Count") 
```
```{r}
#making ggplot histpgram 
fnHC <- as.numeric(jeager2018_counts[jeager.fos.idx, fosneg.HC ])
fn1hr <- as.numeric(jeager2018_counts[jeager.fos.idx, fosneg.1hr])
fp1hr <- as.numeric(jeager2018_counts[jeager.fos.idx, fospos.1hr])

df <- data.frame( fos = as.numeric(fnHC), 
                  cond = factor(rep("Fos- HC", length(fnHC)))
                  )

df2 <- data.frame( fos = as.numeric(fn1hr), 
                  cond = factor(rep("Fos- 1hr", length(fn1hr)))
                  )

df3 <-  data.frame( fos = as.numeric(fp1hr), 
                  cond = factor(rep("Fos+ 1hr", length(fp1hr)))
                  )

df <- rbind(df,df2,df3)

p <- ggplot(df, aes(x=fos, color = cond)) + 
  geom_histogram(fill =  "white") +
  xlab("Fos Reads (seq. depth of 1.5 milion reads per cell)") +
  ylab("Frequency") +
  theme_classic() + 
  theme(legend.title = element_blank()) +
  ggtitle("Freqency of Fos Counts in Jeager Groups")+
  theme(plot.title = element_text(hjust = 0.5))+
  theme(axis.text = element_text(face = "bold")) +
  scale_x_discrete(limits = c(0, 6000)) +
  scale_y_continuous(limits = c(0, 350))

p
```

<br>
So in general there are a few cells with a low level expression of fos, and then this small group of outliers in with a high fos count.  This is in line with the sparse coding of events observed in the brain.  It is notable that within the group that was deemed to be an "engram cell" via antibody staining for fos that there is a not insignificant franction of the fos positive cells which are not expressing fos.  This could be due to off target effects, the authors themselves make note that in CA1 cells they had stained for fos or arc did not express those markers.  It would be better to have a true engram label dataset to work with like Chen et al., (2018) but this si what we had for hippocampus so it will have to do.
<br>
<br>


Looking for Engram labelling gene's in Hochgerner et al., (2018) at P35
---
<br>
<br>
This section is basically just copy pasta from [Seurats tutorials](https://satijalab.org/seurat/articles/pbmc3k_tutorial.html). Here I am using the DEGs published by Jeager et al., (2018).  First we will load the data and the engram cell DEG list, refered to as Jeager DEGs or Jeager Engram DEGs hereafter.  We will stick to the adult DGCs 
<br>
<br>
```{r}
#loading the 5k dataset
# for now I think it is best to keep it seperate from the 24k data as it has only half the genes
# of the 
testsetpath <- "C:/Users/angus/Desktop/test_datasets"

hochgerner5k_2018_counts <- read.table(paste(testsetpath,"/Hochgerner2018/GSE95315_10X_expression_data_v2.tab.gz", sep=""))

colnames(hochgerner5k_2018_counts) <- hochgerner5k_2018_counts[1,]
rownames(hochgerner5k_2018_counts) <- hochgerner5k_2018_counts[,1]

hochgerner5k_2018_meta <- hochgerner5k_2018_counts %>% 
  dplyr::slice(c(1:3)) %>%
  t() %>%
  data.frame %>%
  dplyr::slice(-1) %>% 
  dplyr::select(-cellid)

hochgerner5k_2018_counts <- hochgerner5k_2018_counts %>% 
  dplyr::select(-cellid) %>% 
  dplyr::slice(-c(1:3))


#Load Jeagers DEGs and Check which genes are shared

Jeager.DEGs <- read.csv("Jeager2018_GSE98679/jeager_DEGs_grouped.csv", header = T)

colnames(Jeager.DEGs)[1] <- "GeneID"

DEGs.idx <- Jeager.DEGs$GeneID

DEGs.idx <- DEGs.idx[(DEGs.idx %in% rownames(hochgerner5k_2018_counts))& (DEGs.idx %in% rownames(jeager2018_counts))]

Jeager.DEGs <- Jeager.DEGs[Jeager.DEGs$GeneID %in% DEGs.idx,]

Jeager.DEGs$hoch5k.idx <- which(rownames(hochgerner5k_2018_counts) %in% Jeager.DEGs$GeneID)
```


###  Full gene set in hochgerner
<br>
<br>
First we will cluster within the adult DGCs at post natal day 35 to see if engram cells are aparent just from that.  Although one would expect that some previous group would have noted their existance just from looking for simple cell types.  Previous work on another dataset of DGCs reports no such cell type distinguished by high IEG expression (Erwin et al., 2020; Habib et al., 2016).


```{r}
#get index locations of these gene's in the other datasets
hoch5k.GC_Adult.p35.idx <- (hochgerner5k_2018_meta$age.days.=="35") | (hochgerner5k_2018_meta$age.days.=="35*")
hoch5k.GC_Adult.p35.idx <- (hoch5k.GC_Adult.p35.idx) & (hochgerner5k_2018_meta$cluster_name == "Granule-mature")
hoch5k.GC_Adult.p35.idx <- which(hoch5k.GC_Adult.p35.idx)


#create seurat object just for the Hoch 5k data, with only the JEager DEG counts
#Hoch5k.GCadult <- CreateSeuratObject(counts = hochgerner5k_2018_counts[Jeager.DEGs$hoch5k.idx, hoch5k.GC_Adult.p35.idx]) 

Hoch5k.GCadult <- CreateSeuratObject(counts = hochgerner5k_2018_counts[, hoch5k.GC_Adult.p35.idx])
#this is unesscesary we can do this by changing the, 
#well on second thoguth may be we do need to do this because noramlization may be important to be constrained to just the IEGs
#I think seurat borrows information from other genes. test this
Hoch5k.GCadult
```


### QUALITY CONTROL

Cells and Gene counts
  The dataset provided has already been filtered for cells which, the dataset I refer to as hochgerner 5k is here called dataset A.  It's accession number is GSE95315.  Of the original 6090 cells collected in dataset A they were wilted  down to the 5454 cells in the dataset according to the following measures: "10x Chromium Dataset A (Dataset C) data files were loaded and merged to one dataset. For more uniform representation of timepoints, Dataset C data files were downsampled as indicated in Supplementary Table 5 (‘Cells_analysed’). Valid cells were defined as having >600 (C:>800) total genes, between 800 and 20,000 (C: 1,000–30,000) molecules, and a ratio of molecules to genes >1.2. This resulted in 6,090 (C: 25,546) cells. Next, we removed 46 (C: 732) doublet cells based on coexpression (>1 molecule) of any pair of the following marker genes: Stmn2 (neurons), Mog (oligodendrocytes), Aldoc (astrocytes), C1qc (microglia), and Cldn5 (endothelial). "
 
We can check this by plotting the gene counts and the mitochondria RNA, it is clear that the cells in the dataset have already been subject to the above selection process.  Additionally we will check for over expression of mitochondrial genes as is suggested by the seurate documentation.

```{r}
#make a seurat object of the adult granule cells
Hoch5k.GCadult[["percent.mt"]] <- PercentageFeatureSet(Hoch5k.GCadult, pattern = "^MT-")
```


```{r}
VlnPlot(Hoch5k.GCadult, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


```{r}
plot1 <- FeatureScatter(Hoch5k.GCadult, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Hoch5k.GCadult, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2
```

```{r}

# This is for the filerting out those with features that 
# I'm not sure what the appropriate number for us is, in the tutorial there were 13700 nfeatures
#and the filterd out cells with less then 200 and greater than 2500, so I guess  cells with more than
# a quarter of nFeatures and less than 10% of that is unacceptable?  
Hoch5k.GCadult <- subset(Hoch5k.GCadult, subset = nFeature_RNA > 600 & nFeature_RNA < 20000 & percent.mt < 5)

#note I am going to skip some QC steps here, but we do need to do that in the future
Hoch5k.GCadult <- NormalizeData(Hoch5k.GCadult) # Seurat normalization has many steps see publication
# https://www.frontiersin.org/articles/10.3389/fgene.2020.00041/full describes and compares it to others
Hoch5k.GCadult <- FindVariableFeatures(Hoch5k.GCadult) # Identifies features that are outliers on a ’mean variability plot’.

```

Now we can check the most variable genes which seurat will restrict further analyses to.  Two of the top 10 most variable genes are in the Jeager DEGs, Inhba and Tax1bp1.  

```{r}
top10 <-  head(VariableFeatures(Hoch5k.GCadult), 10)
plot1 <- VariableFeaturePlot(Hoch5k.GCadult)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```

Lets look at the top 100 of the 2000 genes seurat will use and check the timepoint(s) Jeager et al., (2018) assigned them to.  For the most part these DEGs are positively expressed.
```{r}
top100 <-  head(VariableFeatures(Hoch5k.GCadult), 100)
top100[which(top100 %in% Jeager.DEGs$GeneID)]
Jeager.DEGs$group.number[Jeager.DEGs$GeneID %in% top100]
Jeager.DEGs$group[Jeager.DEGs$GeneID %in% top100]
```

```{r}
#scaling for dimnesionality reduction
all.genes <- rownames(Hoch5k.GCadult)
Hoch5k.GCadult <- ScaleData(Hoch5k.GCadult, features = all.genes) 

Hoch5k.GCadult <- RunPCA(Hoch5k.GCadult , features = all.genes,
                         npcs = 20) 
```
<br>
<br>
```{r}
#Dsicovering dimensionality of these activity markers
Hoch5k.GCadult <- JackStraw(Hoch5k.GCadult, num.replicate = 100)
Hoch5k.GCadult <- ScoreJackStraw(Hoch5k.GCadult, dims = 1:20)
```
<br>
<br>
```{r}
JackStrawPlot(Hoch5k.GCadult, dims = 1:20)
```
<br>
<br>
Up to Pc3 we see decently significance, though later we get 4 cluster, though tow of them are barely distinguishable and posses basically no activity dependent markers (Jeager DEGs whihc could serve as markers.
<br>
<br>

```{r}
ElbowPlot(Hoch5k.GCadult)
```
<br>
<br>
Shows more or less the same thing, 4 isn't quite at the elbow but it may as well be.  I played with the QC thresholds and if you increase the number of max nFeatures to high you will reduce to a single PC.
<br>
<br>


### Number of Dimensions affects results

<br>
<br>
I would say that the best cut off would be to go up to dimensions 8, but when this is chosen (by setting dims = 1:8 in FindNeighbors) that doesn't 

```{r}
#Adding metadata to identify the 
Hoch5k.GCadult <- AddMetaData(object = Hoch5k.GCadult, 
                               metadata = hochgerner5k_2018_meta[ hoch5k.GC_Adult.p35.idx, ])

#Clustering, we will visualize in UMP later
Hoch5k.GCadult <- FindNeighbors(Hoch5k.GCadult, dims = 1:3)
Hoch5k.GCadult <- FindClusters(Hoch5k.GCadult, resolution = 0.2)
```


```{r}
Hoch5k.GCadult <- RunUMAP(Hoch5k.GCadult, dims = 1:4)

DimPlot(Hoch5k.GCadult, reduction ="umap")
```

```{r}
DimPlot(Hoch5k.GCadult, reduction ="pca")
```



```{r}
Hoch5k.GCadult.markers.3dims <- FindAllMarkers(Hoch5k.GCadult, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
seurat_clusters.3dims <- Hoch5k.GCadult@meta.data$seurat_clusters
```
```{r}
for (i in c(0:3)){
  print(paste("Cluster ", as.character(i)))
  mrks <- Hoch5k.GCadult.markers.3dims $gene[Hoch5k.GCadult.markers.3dims$cluster==i & Hoch5k.GCadult.markers.3dims$p_val_adj<0.05]
  print(mrks)
  print("Gene in the Jeager DEGs are....")
  print(Jeager.DEGs$GeneID[which(Jeager.DEGs$GeneID %in% mrks)])
  print("And their respective jeager goupings are...")
  print(Jeager.DEGs$group[which(Jeager.DEGs$GeneID %in% mrks)])
  print("")
  print("")
}

```
```{r}
Hoch5k.GCadult.markers.3dims %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
```





```{r}
# trying again with more dimensions
#Clustering, we will visualize in UMP later
Hoch5k.GCadult <- FindNeighbors(Hoch5k.GCadult, dims = 1:7)
Hoch5k.GCadult <- FindClusters(Hoch5k.GCadult, resolution = 0.2)
```


```{r}
Hoch5k.GCadult <- RunUMAP(Hoch5k.GCadult, dims = 1:4)

DimPlot(Hoch5k.GCadult, reduction ="umap")
```

```{r}
DimPlot(Hoch5k.GCadult, reduction ="pca")
```



```{r}
Hoch5k.GCadult.markers.7dims <- FindAllMarkers(Hoch5k.GCadult, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
seurat_clusters.7dims <- Hoch5k.GCadult@meta.data$seurat_clusters
```
```{r}
for (i in c(0:1)){
  print(paste("Cluster ", as.character(i)))
  mrks <- Hoch5k.GCadult.markers.7dims$gene[Hoch5k.GCadult.markers.7dims$cluster==i & Hoch5k.GCadult.markers.7dims$p_val_adj<0.05]
  print(mrks)
  print("Gene in the Jeager DEGs are....")
  print(Jeager.DEGs$GeneID[which(Jeager.DEGs$GeneID %in% mrks)])
  print("And their respective jeager goupings are...")
  print(Jeager.DEGs$group[which(Jeager.DEGs$GeneID %in% mrks)])
  print("")
  print("")
}

```
```{r}
Hoch5k.GCadult.markers.7dims %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
```



```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Fos", "Inhba"))
```
```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Npas4", "Arc"))
```

```{r}
FeaturePlot(Hoch5k.GCadult, features = c("Egr1", "Nrp1"))
```


So I am observing the Penk and Cck seperation that has been described previously (Erwin et al., 2020).  Overall I am not too sure what to make of this analysis.  As it is within a single cell type it may not make sense to try to cluster it further as continuous variation within these cells may be more accurate (Cembrowski & Mannon, 2018). Because there was a sharp and sudden drop in explained variance between PC3 and PC4 so I reduced the clustering to run using only the top 3 PCs as well as the clustering using up to PC7.  If you include up to PC4 you will find the same seperation of the population into 2 clusters, but neither group really has a remarkably higher presence of Jeager DEGs in it.  I thought that prehaps the other dimesnions may be obscuring some more important processes represented by the top 3 PCs.  Running it including only 3 dimensions gives 4 clusters though I question the validity of the first 2.  

It is worth noting the components of PC3 however.

```{r}
VizDimLoadings(Hoch5k.GCadult, dims = 3, reduction = "pca")
```
I would like to point to the presence of Cck as a positive loading and Penk as a negative loading. In Erwin et al., (2020) they
describe a population of mature dentate gyrus cells that are Penk expressing and over represented in the activity labelled cells across a wide variety of hippocampal dependent behaviours during initial training.  Critically in the negative componenets several other Jeager DEGs including Rgs4 and Inhba, in fact Rgs4 is a marker for cluster 1 in the 7 dimensional clustering.  Cck and Penk cells did not overlap, nor do they here as shown in the figure below.  Penk expressing cells are also rare as would be expected.

```{r}
#according to Erwin et al., (2020) Cck and Penk should be mutually exclusively expressed
FeaturePlot(Hoch5k.GCadult, features = c("Penk", "Cck"))
```

Cembroskiw and Mannon (2018) notes that within a single cell type discrete clusters may not always be the best way to think about distiguishing cells.  Continuous cell states such as those underlying neuroplasticity may be at play and confused for distinct cells states.  In this cases it may be useful to just look at the compoenents of PCA and see which ones are most loaded with Jeager DEGs.  As 200 out of 900 of the Jeager DEGs were in the genes included by seurat we can assume 

```{r}
print(Hoch5k.GCadult[["pca"]], dims = 1:7, nfeatures = 50)
```
It may make sense to remove these JeagerDEG genes from the list entirely and try to see how it affects the clustering and cluster just using the DEGs.


* get the cembrowski and mannon  iitation its the one on continuous variation within cell types


Clustering just within the Jeager DEGs
---



```{r}
p <- DimPlot(Hoch5k.GCadult, reduction ="umap")

p <- p + ggtitle("Adult DGCs clustered with Engram DEGs (UMAP)") +
  theme(plot.title = element_text(hjust = 0.5)) +
  xlab("Dim 1") +
  ylab("Dim 2")
  
  
p
# FOR REFERENCE: 
# ggplot(df, aes(x=fos, color = cond)) + 
#   geom_histogram(fill =  "white") +
#   xlab("Fos Reads") +
#   ylab("Frequency") +
#   theme_classic() + 
#   theme(legend.title = element_blank()) +
#   ggtitle("Freqency of Fos Counts in Jeager Groups")+
#   theme(plot.title = element_text(hjust = 0.5))+
#   theme(axis.text = element_text(face = "bold"))
  
```


<br>
<br>
So what we get out of this is some kind of graded expression within the population.  It would be interesting to try to see which cells have more of which gene group in them.  This is similar to what was described by Dr. Cembrowski's group in Erwin et al., (2020) see figure 5 A in particular.  I tried in with 3 dims and I think it looks better with 4 more graded less like swiss cheese.  it would be interesting to see how Jeager's group labels map onto this.
<br>
<br>
```{r}
#Lets find some markers 
cluster0.markers <- FindMarkers(Hoch5k.GCadult, ident.1 = 0, min.pct = 0.25)
head(cluster0.markers, n = 10)
```
```{r}
cluster1.markers <- FindMarkers(Hoch5k.GCadult, ident.1 = 1, min.pct = 0.25)
head(cluster1.markers, n = 10)
```

```{r}
cluster2.markers <- FindMarkers(Hoch5k.GCadult, ident.1 = 2, min.pct = 0.25)
head(cluster2.markers, n = 10)
```

```{r}
cluster3.markers <- FindMarkers(Hoch5k.GCadult, ident.1 = 3, min.pct = 0.25)
head(cluster3.markers, n = 30)
```
```{r}
sum(cluster0.markers$p_val_adj<0.05)
sum(cluster1.markers$p_val_adj<0.05)
sum(cluster2.markers$p_val_adj<0.05)
sum(cluster3.markers$p_val_adj<0.05)

```
```{r}
sd(abs(cluster0.markers$avg_log2FC[cluster0.markers$p_val_adj<0.005]))
sd(abs(cluster1.markers$avg_log2FC[cluster1.markers$p_val_adj<0.005]))
sd(abs(cluster2.markers$avg_log2FC[cluster2.markers$p_val_adj<0.005]))
sd(abs(cluster3.markers$avg_log2FC[cluster3.markers$p_val_adj<0.005]))
```
```{r}
cluster3.markers$avg_log2FC[cluster3.markers$p_val_adj<0.05]

rownames(cluster3.markers)[cluster3.markers$p_val_adj<0.01]
cluster3.markers$avg_log2FC[cluster3.markers$p_val_adj<0.01]
```


<br>
<br>
The only two clusters with really significant are cluster 0 and cluster 3.  Inhba shows significantly elevated expression.  it is by far the most significant marker in the whole dataset.  Inhba is a expressed for a longer time than Arc or Fos according to Jeager et al., (2018) so it makes sense this would be a marker we detect.  Crucially cluster3 cells are also relatively rare compared to the other clusters, an indication of participation in sparse coding.  Cluster 0 has the Melatonin1 receptor has been described previously as being present in dentate gyrus though it is difficult what to make of this.  I foudn a review which suggests Mlt1 is expressed in molecular layer and hilus (the parts of the dentate dyrus that sandwich the granule cell layer), so prehaps we found either a subset of borderline granule cells or they were misclassified during clustering in Hochgerner et al., (2018).  Inhba also tracked Arc expression in the initial stages of engram formation before taking off on it's own as arc dies down between 4 and 5 hours after activation.  Hochgerner also has Gabanergic neurons we could similarily cluster I will begin reviewing the literature on inhibitory interneuron engrams.  Plotting with PCA actually looks even beter than the UMAP plot.
<br>
<br>

```{r}
DimPlot(Hoch5k.GCadult, reduction = "pca")
```
```{r}
p <- DimPlot(Hoch5k.GCadult, reduction = "pca") +
  ggtitle("Hochgerner Adult DGCs clustered with Jeager DEGs") +
  scale_color_discrete(name = "Cluster") + #not scale_fill
  theme_classic() +
  theme(plot.title = element_text(hjust = 0.5)) +
  xlab("PC1") +
  ylab("PC2") 
  

p
```

```{r}
#meh kinda pointless PCA still look the best for illustrating my point
Hoch5k.GCadult <- RunTSNE(Hoch5k.GCadult)

p <- DimPlot(Hoch5k.GCadult, reduction = "tsne") +
  ggtitle("Adult DGCs clustered with Engram DEGs (tSNE)") +
  theme(plot.title = element_text(hjust = 0.5)) 

p
```



<br>
<br>
The other clusters are more or less intermixed while the *Inhba* labelled neurons are clearly moving along PC1.
<br>
<br>

```{r}
Hoch5k.GCadult.markers <- FindAllMarkers(Hoch5k.GCadult, only.pos = TRUE, min.pct = 0.25, logfc.threshold = 0.25)
Hoch5k.GCadult.markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
```
```{r}
Hoch5k.GCadult.markers %>%
    group_by(cluster) %>%
    top_n(n = 10, wt = avg_log2FC) -> top10

DoHeatmap(Hoch5k.GCadult, features = top10$gene, label = FALSE) + 
  theme(axis.text.y = element_text(face = "bold")) +
  theme(legend.title = element_text(face = "bold")) + 
  ggtitle("Hochgerner P35 Mature DCGs Clustered with Jeager Engram DEGs")

```
<br>
<br>


```{r}
clstr3.mrks <- Hoch5k.GCadult.markers[Hoch5k.GCadult.markers$cluster=="3",]

clstr3.mrks$JeagerGrouping <- Jeager.DEGs$group[Jeager.DEGs$GeneID %in% clstr3.mrks$gene]
clstr3.mrks$JeagerGrouping
```
<br>
<br>
Despite 1hr group having the largest portion of the gene's in the Jeager data the majority of the markers here are from the 1hr_4hr_and_5hr group  and the rest are 4hr or 5hr.  I will check their adjusted p-values and the fold change of gene's which cross significance.
<br>
<br>

```{r}
arrange(clstr3.mrks, p_val_adj)
```

```{r}

clstr3.mrks <- Hoch5k.GCadult.markers[Hoch5k.GCadult.markers$cluster=="3",]

clstr3.mrks$JeagerGrouping <- Jeager.DEGs$group[Jeager.DEGs$GeneID %in% clstr3.mrks$gene]

clstr3.mrks[clstr3.mrks$p_val_adj<0.05,] %>%
  arrange( avg_log2FC)
```
```{r}
clstr0.mrks <- Hoch5k.GCadult.markers[Hoch5k.GCadult.markers$cluster=="0",]

clstr0.mrks$JeagerGrouping <- Jeager.DEGs$group[Jeager.DEGs$GeneID %in% clstr0.mrks$gene]

clstr0.mrks[clstr0.mrks$p_val_adj<0.05,] %>%
  arrange( avg_log2FC)

```



<br>
<br>
When I organized them by adjusted pvalues the most significant are the 1hr gene's which I suppose would indicate these cells are recently active.  But thresholding the adjusted p-value for those that cross, we see that 23 gene's survive this thresholding and the gene's with the greatest fold change are almost all the 1hr_4hr_and_5hr genes.  So does this mean these cells are recently active but the genes with the greatest persistence are the ones which undergo the largest gene change?  Alternatively the transient nature of the 1hr markers could be driving down their fold change.  it will be interesting to compare these results to a non-linear classifier.
<br>
<br>
A note on the adjusted p-values from the Seurat reference manual page 74 for the function FindAllMarkers: "p-value adjustment is performed using bonferroni correction based on the total number of genes in
the dataset. Other correction methods are not recommended, as Seurat pre-filters genes using the
arguments above, reducing the number of tests performed. Lastly, as Aaron Lun has pointed out,
p-values should be interpreted cautiously, as the genes used for clustering are the same genes tested
for differential expression."
<br>
<br>

NEXT STEPS
---
<br>
<br>
Cells from Hochgerner et al., (2018) have all passed theri quality controls, I may need to include DGC markers as a sort of control to see if they drive any clustering differently or if Jeager markers are over represented in a particular cell type.
Should also perform this analysis on immature granule cells as well and see if the same markers show up or slightly different ones.rm
<br>
<br>
View this tutorial later:  https://hbctraining.github.io/In-depth-NGS-Data-Analysis-Course/sessionIV/lessons/SC_marker_identification.html

May have good information for labeling FeaturePlot function in seurat with gene expression information.

```{r}
#saving our object and it's data, commented out so its doesn't run everytime

# write.csv(Hoch5k.GCadult.markers, "SeuratAnalysesData/HochgernerGDC_Adult_marker.csv")
# 
# remotes::install_github("mojaveazure/seurat-disk")
# library(SeuratDisk)
# SaveH5Seurat(Hoch5k.GCadult, filename = "SeuratAnalysesData/HochgernerDGC_Adult_P35.h5Seurat", overwrite = TRUE)
```

<br>
<br>

References
---
<br>
<br>
Erwin, S. R., Sun, W., Copeland, M., Lindo, S., Spruston, N., & Cembrowski, M. S. (2020). A sparse, spatially biased subtype of mature granule cell dominates recruitment in hippocampal-associated behaviors. Cell Reports, 31(4), 107551. 
<br>
<br>
Giorgi, C., & Marinelli, S. (2021). Roles and Transcriptional Responses of Inhibitory Neurons in Learning and Memory. Frontiers in Molecular Neuroscience, 113.
<br>
<br>
Jaeger, B. N., Linker, S. B., Parylak, S. L., Barron, J. J., Gallina, I. S., Saavedra, C. D., ... & Gage, F. H. (2018). A novel environment-evoked transcriptional signature predicts reactivity in single dentate granule neurons. Nature communications, 9(1), 1-15.
<br>
<br>
Joglekar, A., Prjibelski, A., Mahfouz, A., Collier, P., Lin, S., Schlusche, A. K., ... & Tilgner, H. U. (2021). A spatially resolved brain region-and cell type-specific isoform atlas of the postnatal mouse brain. Nature Communications, 12(1), 1-16.
<br>
<br>
Habib, N., Li, Y., Heidenreich, M., Swiech, L., Avraham-Davidi, I., Trombetta, J. J., ... & Regev, A. (2016). Div-Seq: Single-nucleus RNA-Seq reveals dynamics of rare adult newborn neurons. Science, 353(6302), 925-928.
<br>
<br>
Hochgerner, H., Zeisel, A., Lönnerberg, P., & Linnarsson, S. (2018). Conserved properties of dentate gyrus neurogenesis across postnatal development revealed by single-cell RNA sequencing. Nature neuroscience, 21(2), 290-299.
<br>
<br>
Klosen, P., Lapmanee, S., Schuster, C., Guardiola, B., Hicks, D., Pevet, P., & Felder‐Schmittbuhl, M. P. (2019). MT1 and MT2 melatonin receptors are expressed in nonoverlapping neuronal populations. Journal of pineal research, 67(1), e12575.
<br>
<br>
Raivich, G., Bohatschek, M., Da Costa, C., Iwata, O., Galiano, M., Hristova, M., ... & Behrens, A. (2004). The AP-1 transcription factor c-Jun is required for efficient axonal regeneration. Neuron, 43(1), 57-67.
<br>
<br>
