---
title: "DG_1to5hoursNE_classifier"
author: "Angus Campbell"
date: "17/03/2022"
output: html_document
---

HOW TO GET THE REACTIVATION SIGNATURE:
-this is not denoted in their metadata
-you must cluster the cells together with the cells from the time course data,
-the cells in condition A>A whihc cluster with the newly active cells are taken to be recently active cells
-the late signature cells are reactivated cells

```{r}
# first attempt at a classifier
setwd("C:/Users/angus/Desktop/PavLabEngrams/EngramCellClassifier")

library(tidyverse)
library(GEOquery)
library(AnnotationDbi)
library(randomForest)
library(rfUtilities) # tutorial here: https://evansmurphy.wixsite.com/evansspatial/random-forest-sdm
library(data.table)
library(reshape2)
library(FactoMineR)
library(factoextra)
library(Rtsne)
library(dplyr)
library(Seurat)
library(stringr)
library(patchwork)
library(metap)
library(cqn)
library(GeoTcgaData)
library(caTools)
# BiocManager::install("cqn")

# Lacar paper: https://www.nature.com/articles/ncomms11022?origin=ppub
# Jeager paper: https://www.nature.com/articles/s41467-018-05418-8

```


```{r}
# Lacar et al., (2016)
lacar2016_meta <- read.csv('Lacar2016_GSE77067/SraRunTable.txt', header = TRUE)
lacar2016_snHC_counts <- read.table('Lacar2016_GSE77067/GSE77067_sn_hc_counts.txt.gz')
lacar2016_snNE_counts <- read.table('Lacar2016_GSE77067/GSE77067_sn_ne_counts.txt.gz')
lacar2016_wc_counts <- read.table('Lacar2016_GSE77067/GSE77067_wc_counts.txt.gz')


```




```{r}
#Jeagers meta rows are a little out of order wrt their counts, i.e. rows do no correspond to cells order we fix that in a bit
jeager2018_counts <- bind_cols(read.table('Jeager2018_GSE98679/GSE98679_count.txt.gz', header = TRUE, check.names = FALSE),
                               read.table('Jeager2018_GSE98679/GSE98679_v2_GSM3308862-GSM3309413_count.txt.gz', header = TRUE, check.names = FALSE))

jeager2018_meta <- read.csv('Jeager2018_GSE98679/SraRunTable.txt', header = TRUE)
jeager2018_meta = jeager2018_meta[c(1:46,599:912,47:598),] #here we fix the order
rownames(jeager2018_meta) <- c(1:912)

jeager2018_meta$predicted_cell_type <- as.character(lapply(jeager2018_meta$predicted_cell_type, function(x) if (x=="") {"DG"} else {x}))
jeager2018_meta$predicted_cell_type <- lapply(jeager2018_meta$predicted_cell_type, function(x) if (x=="") {"DG"} else {x})

jeager2018_meta$fos_status <- as.factor(sapply(as.character(jeager2018_meta$source_name), function(y) if (grepl("_F_", y, fixed=TRUE)) "Fos+" else "Fos-"  ))
jeager2018_meta$fos_status[361:dim(jeager2018_meta)[1]] <- "Fos+"

jeager2018_meta$Mouse_Number[c(361:912)] <- jeager2018_meta$mousingle_number[c(361:912)]
jeager2018_meta <- jeager2018_meta %>% 
  dplyr::select(-mousingle_number)

#filtering out cells as per instructions in Jeager et al., 2018)
under4k <- sapply(jeager2018_counts, function(y) sum(length(which(y>0))))

# as per the mthods section we remove cells with less than 4k gene expressed or 100k reads aligned
filtered.idx <- as.numeric(which((under4k>4000)&(jeager2018_meta$alignable>100000)))
filtered.idx <- order(c(filtered.idx,194))

jeager2018_counts <-jeager2018_counts[,filtered.idx]
jeager2018_counts[is.na(jeager2018_counts)] <- 0

jeager2018_meta <- jeager2018_meta[filtered.idx,]




```


```{r}
#matching all genes and merging datasets
multi.intersect <- function(x) Reduce(intersect, x) #takes lists of lists, c() will not work

shared.genes <- multi.intersect(list(rownames(jeager2018_counts),
                           rownames(lacar2016_wc_counts),
                           rownames(lacar2016_snHC_counts),
                           rownames(lacar2016_snNE_counts)
                          )
                        )

jeager2018_counts$gene <- as.character(rownames(jeager2018_counts))
jeager2018_counts <- jeager2018_counts[rownames(jeager2018_counts) %in% shared.genes,]
#jeager2018_counts <- lastcol.to.firstcol(jeager2018_counts)

lacar2016_wc_counts$gene <- as.character(rownames(lacar2016_wc_counts))
lacar2016_wc_counts <- lacar2016_wc_counts[rownames(lacar2016_wc_counts) %in% shared.genes,]

#lacar2016_wc_counts <- lastcol.to.firstcol(lacar2016_wc_counts)

lacar2016_snHC_counts$gene <- as.character(rownames(lacar2016_snHC_counts))
lacar2016_snHC_counts <-lacar2016_snHC_counts[rownames(lacar2016_snHC_counts) %in% shared.genes,]
#lacar2016_snHC_counts <- lastcol.to.firstcol(lacar2016_snHC_counts)

lacar2016_snNE_counts$gene <- as.character(rownames(lacar2016_snNE_counts))
lacar2016_snNE_counts[rownames(lacar2016_snNE_counts) %in% shared.genes,]
#lacar2016_snNE_counts <- lastcol.to.firstcol(lacar2016_snNE_counts)

# we will remove the PTZ treated cells as well before matching its genes
not.ptz <- which(lacar2016_meta$treatment != "PTZ")


#Match the gene sets by adding the gene names as rows, will strip later
DG.idx <- which(jeager2018_meta$predicted_cell_type=="DG")

# we must add 1 to the values of DG.idx and not.ptz to deal with the generow shifting the index 1
combined.counts <- jeager2018_counts[, c(DG.idx,862)] %>% 
  left_join(lacar2016_wc_counts[, c(not.ptz[not.ptz <= 82],83)], by = 'gene' , all.y = TRUE) %>%
  left_join(lacar2016_snHC_counts, by = 'gene', all.y = TRUE) %>%
  left_join(lacar2016_snNE_counts, by = 'gene', all.y = TRUE) #%>% 
  #tibble::column_to_rownames(var="gene") %>% this throws an error, it would be good
  #dplyr::select(-gene)

#this join is possibly inclduing the gene rows leading to mismathc number of cells later 

#give the combined.counts genes for rownames and get rid of that column
rownames(combined.counts) <- combined.counts$gene
combined.counts$gene <- NULL
combined.counts[is.na(combined.counts)] <- 0

#cleaning up the gene column from the other count data
jeager2018_counts$gene <- NULL
lacar2016_wc_counts$gene <- NULL
lacar2016_snHC_counts$gene <- NULL
lacar2016_snNE_counts$gene <- NULL

```



```{r}

#modify this line in order to change the strings in the metadata below
#jeager2018_meta$predicted_cell_type <- as.character(lapply(jeager2018_meta$predicted_cell_type, function(x) if (x=="") {"DG"} else {x}))


#columns for which paper the cells are from
experiment.label <- c(replicate( length(DG.idx),"Jeager" ),
                 replicate( length(not.ptz),"Lacar" ))


#column for treatment
treatment <- c(jeager2018_meta$exposure[DG.idx],
               lacar2016_meta$treatment[not.ptz])
treatment <- as.character(lapply(treatment, function(x) if (x=="home-cage") {"HC"} else {x}))
treatment <- as.character(lapply(treatment, function(x) if (x=="novel environment") {"NE"} else {x}))

#fos status 
fos_status <-c(as.character(jeager2018_meta$fos_status[DG.idx]),
               lacar2016_meta$facs_sort[not.ptz])
fos_status <- as.character(lapply(fos_status, function(x) if (x=="Prox1+/Fos+") {"Fos+"} else {x}))
fos_status <- as.character(lapply(fos_status, function(x) if (x=="NeuN+/Prox1+") {"Fos-"} else {x}))
fos_status <- as.character(lapply(fos_status, function(x) if (x=="Prox1+/Fos-") {"Fos-"} else {x}))
fos_status <- as.character(lapply(fos_status, function(x) if (x=="GFP+") {"Fos-"} else {x}))
# need to check what GFP + means, I think it means it is fos positive


combined.meta <- data.frame(experiment.label,
                 treatment,
                 fos_status)
#this throws an error mismathc number of rows and genes most likely
rownames(combined.meta) <- colnames(combined.counts)
```


L3 <- LETTERS[1:3]
fac <- sample(L3, 10, replace = TRUE)
(d <- data.frame(x = 1, y = 1:10, fac = fac))
## The "same" with automatic column names:
data.frame(1, 1:10, sample(L3, 10, replace = TRUE))






###THings to include in our model


1) test if rank transform or onehotencoding is best, may be one with both

2) try including different datasets all at once

3) try including a term for cell type, could include a marker set so the classifier could easily do this itself,
throws out unclassified cells or classifies them seperately or try just 

4) Try to include epilepsy cells (PTZ treated from Lacar et al., 2016), according to Jeager these should have fairly distinct transcription profiles.



One hot encoding this
---
```{r}
#no need to scale we are one hot encoding this
onehot <- function(x){
  x <- x<1
  return(as.integer(x))
}


binarized.counts <- data.frame( lapply(combined.counts, function(x) as.character(as.integer(x>1))) )
binarized.counts <- data.frame( t(binarized.counts), stringsAsFactors = TRUE )
binarized.counts$Engramcell <- as.factor(combined.meta$fos_status)

# sample slit comes from library(caTools)

split <- sample.split(binarized.counts$Engramcell, SplitRatio = 0.8)

training_set = subset(binarized.counts, split == TRUE)
test_set = subset(binarized.counts, split == FALSE)

classifier = randomForest(x = training_set[-1],
                          y = training_set$Engramcell,
                          ntree = 500)

#this function may handle all the splitting
# from rfUtilities package
rf.cv <- rf.crossValidation(classifier, training_set[-1], 
                            normalize = FALSE, p=0.125, 
                            n=10, ntree=501)
# plotting
plot(rf.cv, type = "cv", main = "CV producers accuracy") 
plot(rf.cv, type = "model", main = "Model producers accuracy")
plot(rf.cv, type = "cv", stat = "oob", main = "CV oob error")
plot(rf.cv, type = "model", stat = "oob", main = "Model oob error")

# Predicting the Test set results
y_pred = predict(classifier, newdata = test_set)

classifier$importance
importancerank <- data.frame(rownames(combined.counts),
                             classifier$importance)
test <- arrange(importancerank, desc(MeanDecreaseGini))


cv.classifier <- rfcv(trainx = training_set[-1], trainy = training_set$Engramcell,
     cv.fold=5, scale="log", step=0.5)




#split the data
#making the meta data,
# https://rdrr.io/rforge/CrossValidate/man/balancedSplit.html
library(CrossValidate) # use this library to make balanced test trainng and folds of data
library(randomForest)
library(caret) # also has functions to split the data and 
library(rfUtilities)
#  https://cran.r-project.org/web/packages/rfUtilities/rfUtilities.pdf



#
RF.fit = randomForest(x = dat[,c(1:n)], #reisudals from our models
                           y = dat$cbcltotal,  #
                           ntree = treenumber, importance = T)


```



---Looking for Reactivation


Ok so we will need to filter the counts data later for the genes that match in Hochgerner or other datasets we use.
First we will try to reproduce the results of Jeager et al. (2018) in figure 5.


```{r}
#granule cells ,DG whihc were not in the homecage condition HC
DG.idx <- which(jeager2018_meta$predicted_cell_type=="DG")


rownames(jeager2018_meta) <- colnames(jeager2018_counts)
Jeager2018 <- CreateSeuratObject(counts = jeager2018_counts[,DG.idx],
                                 meta.data = jeager2018_meta[DG.idx,])

#this is the only filtering step I will retain from seurat as we have already done filterin earlier
# based on reads aligned and total gene expressed in the last section
Jeager2018[["percent.mt"]] <- PercentageFeatureSet(Jeager2018, pattern = "^MT-")

# this doesnt tell you much but its ok, looks ugly but yes basically there are no unacceptablely hihg or low expressiion
# and mitochondrial genes are in acceptable ranges
VlnPlot(Jeager2018, features = c("nFeature_RNA", "nCount_RNA", "percent.mt"), ncol = 3)
```


```{r}

plot1 <- FeatureScatter(Jeager2018, feature1 = "nCount_RNA", feature2 = "percent.mt")
plot2 <- FeatureScatter(Jeager2018, feature1 = "nCount_RNA", feature2 = "nFeature_RNA")
plot1 + plot2

```


```{r}

Jeager2018 <- NormalizeData(Jeager2018, normalization.method = "LogNormalize", scale.factor = 10000)

Jeager2018 <- FindVariableFeatures(Jeager2018, selection.method = "vst", nfeatures = 8000)

# Identify the 10 most highly variable genes
top10 <- head(VariableFeatures(Jeager2018), 10)

# plot variable features with and without labels
plot1 <- VariableFeaturePlot(Jeager2018)
plot2 <- LabelPoints(plot = plot1, points = top10, repel = TRUE)
plot2
```


```{r}
sum(jeager2018_meta$fos_status=="Fos+")

all.genes.jeager <- rownames(Jeager2018)
Jeager2018 <- ScaleData(Jeager2018, features = all.genes.jeager)

```


```{r}
Jeager2018 <- RunPCA(Jeager2018, 
                     features = VariableFeatures(object = Jeager2018),
                     npcs = 50)


```


```{r}

Jeager2018 <- JackStraw(Jeager2018, num.replicate = 100)
Jeager2018 <- ScoreJackStraw(Jeager2018, dims = 1:20)

JackStrawPlot(Jeager2018)
```

```{r}
ElbowPlot(Jeager2018)
```


```{r}
#Clustering, we will visualize in UMP later
Jeager2018 <- FindNeighbors(Jeager2018, dims = 1:6)
Jeager2018 <- FindClusters(Jeager2018, resolution = 0.2) #resolution of 0.5 finds 4 clusters, 0.2 combines all the late signature into one cluster, are these rpehaps reactivated cells?  Or just really active cells
#PCA makes it seem like cells are definitely moving from baseline into distinct activation paths
Jeager2018@meta.data$PointTwoClusters <- Jeager2018@meta.data$seurat_clusters

Jeager2018 <- FindClusters(Jeager2018, resolution = 0.5) 
Jeager2018@meta.data$PointFiveClusters <- Jeager2018@meta.data$seurat_clusters
```

```{r}
Jeager2018 <- RunTSNE(Jeager2018, ndims = 1:6)
DimPlot(Jeager2018, reduction = 'tsne', group.by = "exposure") + 
  DimPlot(Jeager2018, reduction = 'tsne')
```

```{r}
#look at pca
DimPlot(Jeager2018, reduction = 'pca', group.by = "PointTwoClusters") + 
  DimPlot(Jeager2018, reduction = 'pca', group.by ='exposure')
```
```{r}
DimPlot(Jeager2018, reduction = 'pca', group.by = "PointFiveClusters") +
  DimPlot(Jeager2018, reduction = 'pca', group.by ='exposure')

```

```{r}
DimPlot(Jeager2018, reduction = 'pca', group.by = "fos_status") +
  DimPlot(Jeager2018, reduction = 'pca', group.by ='exposure')
```



```{r}
jeager.markers <- FindAllMarkers(Jeager2018, only.pos = FALSE, min.pct = 0.25, logfc.threshold = 0.25)
jeager.markers %>%
    group_by(cluster) %>%
    slice_max(n = 2, order_by = avg_log2FC)
```

```{r}
jeager.markers
```



```{r}
Jeager2018 <- RunUMAP(Jeager2018, dims = 1:4)
DimPlot(Jeager2018, reduction = 'pca')

```



```{r}
#https://www.rdocumentation.org/packages/Seurat/versions/2.3.4/topics/BuildRFClassifier
BuildRFClassifier(Jeager2018, training.genes = NULL, training.classes = NULL,
  verbose = TRUE, ...)

# Following this we can run the data through this function
ClassifyCells(jeagerdata, classifier, training.genes = NULL,
              training.classes = NULL, new.data = test.set, ...)

```


#Ideas for classifiers

1) Simply which cells have a bunch of known IEGs highly expressed?

2) Fold change then relu


This paper needs to be looked at and readFernandez-Albert, J., Lipinski, M., Lopez-Cascales, M. T., Rowley, M. J., Martin-Gonzalez, A. M., Del Blanco, B., ... & Barco, A. (2019). Immediate and deferred epigenomic signatures of in vivo neuronal activation in mouse hippocampus. Nature neuroscience, 22(10), 1718-1730.
https://www.nature.com/articles/s41593-019-0476-2#data-availability
Data:  https://www.ncbi.nlm.nih.gov/geo/query/acc.cgi?acc=GSE125068



#PLaying around looking for Penk

Again ti is incredibly lowly expressed in general, a seubset of cells

```{r}
# which(rownames(jeager2018_counts) == "Penk")
# 34282
# > sum((jeager2018_counts[34282,DG.idx]>1)&(jeager2018_meta$fos_status[DG.idx]=="Fos+"))
# [1] 96
# > sum((jeager2018_counts[34282,DG.idx]>1)&(jeager2018_meta$fos_status[DG.idx]=="Fos-"))
# [1] 11
528
sum(jeager2018_meta$fos_status[DG.idx]=="Fos-")

#summaries of...
#Penk expression in Fox- DGCs
summary(as.numeric(jeager2018_counts[34282,DG.idx][which(jeager2018_meta$fos_status[DG.idx]=="Fos-")]))
#Penk expression in Fos+ DGCs
summary(as.numeric(jeager2018_counts[34282,DG.idx][which(jeager2018_meta$fos_status[DG.idx]=="Fos+")]))

#give you quantiles of Penk first in Fos+ cells then in population as a whole
quantile(as.numeric(jeager2018_counts[34282,DG.idx][which(jeager2018_meta$fos_status[DG.idx]=="Fos+")]), prob =0.99)
quantile(as.numeric(jeager2018_counts[34282,DG.idx]), prob =0.8)



sum((jeager2018_counts[34282,DG.idx]>105)&(jeager2018_meta$fos_status[DG.idx]=="Fos-"))
sum((jeager2018_counts[34282,DG.idx]>105)&(jeager2018_meta$fos_status[DG.idx]=="Fos+"))


sum((jeager2018_counts[34282,DG.idx]>1)&(jeager2018_meta$fos_status[DG.idx]=="Fos+")&(jeager2018_meta$exposure[DG.idx]=="1hr"))

sum((jeager2018_counts[34282,DG.idx]>1)&(jeager2018_meta$fos_status[DG.idx]=="Fos+")&(jeager2018_meta$exposure[DG.idx]=="4hr"))

sum((jeager2018_counts[34282,DG.idx]>1)&(jeager2018_meta$fos_status[DG.idx]=="Fos+")&(jeager2018_meta$exposure[DG.idx]=="5hr"))

sum(as.numeric((jeager2018_counts[34282,DG.idx]>1)&(jeager2018_meta$fos_status[DG.idx]=="Fos+")&(jeager2018_meta$exposure[DG.idx]=="A>A")))

sum(as.numeric((jeager2018_counts[34282,DG.idx]>1)&(jeager2018_meta$fos_status[DG.idx]=="Fos+")&(jeager2018_meta$exposure[DG.idx]=="A>C")))




sum(jeager2018_meta$fos_status[DG.idx]=="Fos+")
sum((jeager2018_counts[34282,DG.idx]>105)&(jeager2018_meta$fos_status[DG.idx]=="Fos+"))
sum(jeager2018_meta$fos_status[DG.idx]=="Fos-")
sum((jeager2018_counts[34282,DG.idx]>105)&(jeager2018_meta$fos_status[DG.idx]=="Fos-"))
```
Penk expression is heavily skewed, about a fifth fo cells on this hihg depth sequecning (albeit single nucleus) have more than 1 read count.  At baseline this is about 20% of adult DGCs (Fos-) but only ~10 % express it in the 90th quantile at >105
Penk is very rarely expressed a threshold of 1 read is sufficient to generate a similar result to the on below.  A pattern clearly emerges, it become more obvious as you raise the threshold of Penk.  In general baseline fos- neurons exhibit the same frequency of penk expression as the 5hr group and the 1hr group.  But at 1 and 4 hours there is distinctly lower penk expression, this is espescially noticable at the highest threshold.  At the 99th percentile 9230 reads only the A>A group exhibits penk.
```{r}
#give you quantiles of Penk first in Fos+ cells then in population as a whole
quantile(as.numeric(jeager2018_counts[34282,DG.idx][which(jeager2018_meta$fos_status[DG.idx]=="Fos+")]), prob =0.8)
quantile(as.numeric(jeager2018_counts[34282,DG.idx]), prob =0.8)
# gives you penkthresh

#Plotting Penk expression by exposure
penkfreq <- c()
exposure.list = c("1hr", "4hr", "5hr", "A>C", "A>A")
penk.thresh = 9253 # 80th qyunsatile at 1 read, 90th at 105 and 95th at 1037.8 call it 10 OF Fof+ CELLS not the whol population

#for loop to make test to get frequency of penk expression, 105 was chosen as it represents the top 90th percentile of penk expression, most cells do not express it at all and where it is expressed it is generally highly expresed
for(i in c(1:5)){
  
penk.count <- sum((jeager2018_counts[34282,DG.idx]>penk.thresh)&(jeager2018_meta$fos_status[DG.idx]=="Fos+")&(jeager2018_meta$exposure[DG.idx]==exposure.list[i]))

totalpos <- sum((jeager2018_meta$fos_status[DG.idx]=="Fos+")&(jeager2018_meta$exposure[DG.idx]=="5hr"))

penkfreq <- c(penkfreq , (penk.count/totalpos)*100)

}

test <- data.frame(penkfreq )
test$exposure <- exposure.list

#frequency in fos- adult DGCs
sum((jeager2018_counts[34282,DG.idx]>penk.thresh)&(jeager2018_meta$fos_status[DG.idx]=="Fos-"))/sum(jeager2018_meta$fos_status[DG.idx]=="Fos-")

test

# Results after making test with above for loop
# > test
#     penkfreq exposure
# 1 0.01041667      1hr
# 2 0.02083333      4hr
# 3 0.12500000      5hr
# 4 0.07291667      A>C
# 5 0.32291667      A>A
# > sum((jeager2018_counts[34282,DG.idx]>1)&(jeager2018_meta$fos_status[DG.idx]=="Fos-"))/sum(jeager2018_meta$fos_status[DG.idx]=="Fos-")
# [1] 0.1964286
# > sum((jeager2018_counts[34282,DG.idx]>105)&(jeager2018_meta$fos_status[DG.idx]=="Fos-"))/sum(jeager2018_meta$fos_status[DG.idx]=="Fos-")
# [1] 0.1071429
# > sum((jeager2018_counts[34282,DG.idx]>penk.thresh)&(jeager2018_meta$fos_status[DG.idx]=="Fos-"))/sum(jeager2018_meta$fos_status[DG.idx]=="Fos-")
# [1] 0.1071429


```






WARNING PUT ON YOUR TINFOIL HAT
Penk codes for a Proenkephalin, an opiod receptor.  Exercise is a known inhibitor of memory formation in rodents.  Endogenous opioids increase during exercise (find a reference I would assume they do), and often metabolic pathways are driven down by the excessive presence of a substance.  I.e. think of lactase (may be not the best example), Busckaski showed that blood sugar is driven down by short wave ripples in hippocampus via a hippocampal-hypothalamic pathway.  SWRs tend to occur while mice are standing still, the hippocampus is deeply tied to the animals decision to move and the biological processes .  It is possible that during highly exertive exploration mice are designed not to remeber the event in the long term, to prevent them from returning to the same area.  It will be interesting to see if this is conserved in humans as we seem to be making memories all the time.  Mice tend to replay areas that have been highly rewarded during replay as well this could be a mechanism for enforcing a levy walk between highly rewarding areas while avoiding areas invovling hihgly strenuous activity.  Given teh spatial clustering of penk cells I feel they would have similar connectivity.  It will be interesting to see what features the random forest pulls up or whether exploring these gradient markers may be informative instead.

Opposing theories of exercise focus on neurogenesis, hypothesizing that neurogenesis drives integration of immature granule cells into circuits and disrupt their formation.  This may be possible but I would like to advance this opposite hypothesis here that exercise dampens penk expression which is tied to other more directly memory related pathways as well,  this hypothesis could be tested via flourescent endoenkenephalins probes and a activity label for penk (with a viral construct), to establish if penk expression is diminished by endoenkenephalins.  Or alternatively by washing in interfereing mRNA for Penk during prior to an animal learning and seeing if this has the same level and kinds of deficits as a control animal.
In Melo et al., (2014) Penk mice showed reduced effects of a chronic mild stress protocol.  The authors offered the explanation that penk is therefor a regulator of chronic stress.  But Erwin et al. (2018) found penk expression to be I would suggest instead that Penk in the dentate gyrus along with other Enkephalin related gene's are barometrs of exertion.  These may be down
regulated by enkephalin relese in response to exercise.  They may appear generally across many behaviours there for to serve as a barometer of exertion associated with this sequence of events.  Behaviours that are not strenuous would drive expression of engram cells.  In other words they are a mechanism by which animals tend to remeber and therefor prefer less strenuous environments or choices that are less strenuous to perform.  Penk KO mice are not simply less responsive to stress they are not remembering the events.  A conditioned place preference for nicotine was also absent in Penk KO mice (Berrendero et al., 2005), I would suggest this is due to them being unable to form engram cells representing the region they recieved the reward.  As Berrendo et al., (2005) did not test the performance of the Penk KO mice on a general spatial test their results can not rule this possibility out.

There also appears to be cross talk between circuits as evidenced by generalization such as stress and again this appears to be located ni the dorsal dentate gyrus(Ren et al., 2022).  There appear to be distinct transcriptional responses in different cells within the same memory whihc drive different synaptic propoerties (Sun et al., 2020).  Although the authors of sun et al. (2020) do note that their construct expressed using enhancers for the different IEGs and that the initial IEGs themselves may be general but induce different expression in different populations (this is a question we can answer or look at in literature such as Marco or Fernandez ATAC seq data).  There is an argument in there where they discuss the "generalizability of  memory engrams" the push and pull of engram trace but no discussion of why there would be different memory engrams in the first place.


Denny, C. A., Kheirbek, M. A., Alba, E. L., Tanaka, K. F., Brachman, R. A., Laughman, K. B., ... & Hen, R. (2014). Hippocampal memory traces are differentially modulated by experience, time, and adult neurogenesis. Neuron, 83(1), 189-201.

Sun, X., Bernstein, M. J., Meng, M., Rao, S., Sørensen, A. T., Yao, L., ... & Lin, Y. (2020). Functionally distinct neuronal ensembles within the memory engram. Cell, 181(2), 410-423.

Ren, L. Y., Meyer, M. A., Grayson, V. S., Gao, P., Guedea, A. L., & Radulovic, J. (2022). Stress-induced generalization of negative memories is mediated by an extended hippocampal circuit. Neuropsychopharmacology, 47(2), 516-523.

Melo, I., Drews, E., Zimmer, A., & Bilkei‐Gorzo, A. (2014). Enkephalin knockout male mice are resistant to chronic mild stress. Genes, Brain and Behavior, 13(6), 550-558.
https://pubmed.ncbi.nlm.nih.gov/24804898/#:~:text=Previous%20studies%20using%20preproenkephalin%20knockout,stress%20conditions%20is%20not%20known.

Berrendero, F., Mendizábal, V., Robledo, P., Galeote, L., Bilkei-Gorzo, A., Zimmer, A., & Maldonado, R. (2005). Nicotine-induced antinociception, rewarding effects, and physical dependence are decreased in mice lacking the preproenkephalin gene. Journal of Neuroscience, 25(5), 1103-1112.


Ziółkowska, B., Przewłocka, B., Mika, J., Łabuz, D., & Przewłocki, R. (1998). Evidence for Fos involvement in the regulation of proenkephalin and prodynorphin gene expression in the rat hippocampus. Molecular brain research, 54(2), 243-251.
https://www.sciencedirect.com/science/article/pii/S0169328X97003409?casa_token=Nhia_qNzNzcAAAAA:KK_uoqv-r7vtykVOguDRrkIcJqMO_iveVnafd7O6DQtUUsS-1LC-Xlncp3T1-tNbgnkGQS02mmI

Great review on this topic:
https://www.sciencedirect.com/science/article/pii/S0079612307630155?casa_token=pNgh9WgXC1wAAAAA:S3Zv6zw4xaaGNL4g1o7HBp7LUvt_DY-yUL98bJSPM0CwHOGHY8rgxE2QcdPPbNrC3q9zdiBG0So#fig1

Johnston, H. M., & Morris, B. J. (1994). Induction of c-fos gene expression is not responsible for increased proenkephalin mRNA levels in the hippocampal dentate gyrus following NMDA stimulation. Molecular brain research, 25(1-2), 147-150.


Johnson, M. A., Contoreggi, N. H., Kogan, J. F., Bryson, M., Rubin, B. R., Gray, J. D., ... & Milner, T. A. (2021). Chronic stress differentially alters mRNA expression of opioid peptides and receptors in the dorsal hippocampus of female and male rats. Journal of Comparative Neurology, 529(10), 2636-2657.

Watson, Charles, and Daniel Binks. "Elongation of the CA1 field of the septal hippocampus in ungulates." Journal of Comparative Neurology 527, no. 4 (2019): 818-832.
-shows penk expression in hippocampus is conserved across mammals though it is lower in primates compared to rodents


"Overall, the DG exhibits low levels of activity under normal physiological conditions and
behavior states regardless of whether the animal is immobile or engaging in active behavior,
such as, during exploration and mobility. Only about one to three per-cent of principal granule
cells are active while the majority of the cells remain inactive, and the cells fire only one to two
times in 10 seconds 1,2,6–8"
https://assets.researchsquare.com/files/rs-123425/v1_covered.pdf?c=1631849209
Lipponen, A., Sahramäki, J., Penttonen, M., & Nokia, M. (2020). Dentate gyrus single cell neuronal activity in rats during fixed interval task and subsequent sleep.

Diets can act to decrease neuronal excitability:
Bough, K. J., Schwartzkroin, P. A., & Rho, J. M. (2003). Calorie restriction and ketogenic diet diminish neuronal excitability in rat dentate gyrus in vivo. Epilepsia, 44(6), 752-760.

https://sci-hub.hkvisa.net/10.1080/1028415x.2000.11747302


Bonthuis, P. J., Steinwand, S., Hörndli, C. N. S., Emery, J., Huang, W. C., Kravitz, S., ... & Gregg, C. (2022). Noncanonical genomic imprinting in the monoamine system determines naturalistic foraging and brain-adrenal axis functions. Cell Reports, 38(10), 110500.
-interesting paper

Padilla, S. L., Qiu, J., Soden, M. E., Sanz, E., Nestor, C. C., Barker, F. D., ... & Palmiter, R. D. (2016). Agouti-related peptide neural circuits mediate adaptive behaviors in the starved state. Nature neuroscience, 19(5), 734-741.
-really interesting introduction about food and behaviours
-makes note that in starvation animals become less sensitive to stress and more risk taking.

Uemura, M., Blankvoort, S., Tok, S. S. L., Yuan, L., Cobar, L. F., Lit, K. K., & Tashiro, A. (2021). A neurogenic microenvironment defined by excitatory-inhibitory neuronal circuits in adult dentate gyrus. Cell Reports, 36(1), 109324.
-shows local cell activity can mediate immature cells to integrate into circuits
