---
title: "Reporducing Jeager et al. (2018)"
author: "Angus Campbell"
date: "16/01/2022"
output: html_document
---

https://github.com/adam-p/markdown-here/wiki/Markdown-Cheatsheet

```{r message=FALSE, include=FALSE}
setwd("C:/Users/angus/Desktop/PavLabEngrams/EngramCellClassifier")
library(tidyverse)
library(GEOquery)
library(AnnotationDbi)
library(randomForest)
library(data.table)
library(reshape2)
library(FactoMineR)
library(factoextra)
library(Rtsne)
```

INTORDUCTION
---
  We will attempt to reproduce some of the analyses in [Jeager et al., 2018](https://www.ncbi.nlm.nih.gov/pmc/articles/PMC6079101/).  We are of course espescially interested in the classifier in figure 6C with the 
  
  
LOADING THE DATA
---

```{r}
# Jeager et al., (2018) meta data and tpm

jeager2018_tpm <- bind_cols(read.table('Jeager2018_GSE98679/GSE98679_tpm.txt.gz', header = TRUE, check.names = FALSE),
                               read.table('Jeager2018_GSE98679/GSE98679_v2_GSM3308862-GSM3309413_tpm.txt.gz', header = TRUE, check.names = FALSE))

#We are using the tpms provided to skip some of the filtering steps, but if desreed here's where they are.
#There are 2581 genes excluded after filtering.
#
jeager2018_counts <- bind_cols(read.table('Jeager2018_GSE98679/GSE98679_count.txt.gz', header = TRUE, check.names = FALSE),
                               read.table('Jeager2018_GSE98679/GSE98679_v2_GSM3308862-GSM3309413_count.txt.gz', header = TRUE, check.names = FALSE))

```

```{r}
jeager2018_meta <- read.csv('Jeager2018_GSE98679/SraRunTable.txt', header = TRUE)
jeager2018_meta = jeager2018_meta[c(1:46,599:912,47:598),] #we need to fix Jeager's files up a bit
rownames(jeager2018_meta) <- c(1:912)

# blank.to.DG <-function(x){
#   if(x==""){
#     x="DG"
#   }
# }
jeager2018_meta$predicted_cell_type <- as.character(lapply(jeager2018_meta$predicted_cell_type, function(x) if (x=="") {"DG"} else {x}))

jeager2018_meta$predicted_cell_type <- lapply(jeager2018_meta$predicted_cell_type, function(x) if (x=="") {"DG"} else {x})

#ditch mousingle_number, put it into Mouse_Number
jeager2018_meta$Mouse_Number[361:912] <- jeager2018_meta$mousingle_number[361:912]
jeager2018_meta <- jeager2018_meta[,c(1:33,35,36)]
```

Active DG neurons exhibit a dramatic shift in transcription
---

The thresholding for this first session is described as "Nuclei were excluded as outliers based on total aligned reads and total gene count, or extreme-outliers, based on clustering (Supplementary Fig 2a–b). An average of 1.17 million reads were aligned per nucleus, with an average of 5637 genes detected above a log2(TPM+1) (TPM) = 1."

```{r}
meangenecounts <- function(tpm.df){
  test <- log(tpm.df+1, 2)>1 #log base 2 tpm counts +1 over 1
  return(mean(as.numeric(colSums(test, na.rm = TRUE)))) 
}

print("Average gene's passing threshold per cell:")
print(meangenecounts(jeager2018_tpm[,c(1:360)])) 
print("Total missing entries in first dataset:")
print(sum(is.na(jeager2018_tpm[,c(1:360)]))) 
```

I know some of the nuclei in the data set were considered outliers.  Maybe this is part of the discrepancy?  After all 912 nuclei are included in the data but only 868 were used as per the methods section  "The final cutoffs were 100,000 total aligned reads and 4000 total genes detected. Nuclei below these thresholds were detected as outliers and removed from further analysis. A total of 967 nuclei were used for this study, with 868 (89.8%) surviving the filtering thresholds".

> sum(jeager2018_meta$alignable[c(1:360)]>100000)
[1] 346

> as.numeric(which(test2<4000))
 [1]  49  83 123 137 139 199 200 214 215 221 249 254 256 257 261 267 268 269 270 271
 
We can see which do not have sufficient reads to be included.  Only 194 should be dropped as well base don that criteria.
 
> which(jeager2018_meta$alignable[c(1:360)]<100000)
 [1]  49  83 123 194 199 214 215 221 249 257 261 267 268 269

```{r}
test <- sapply(jeager2018_tpm[,c(1:360)], function(y) sum(length(which(is.na(y)))))
print(summary(test))
```

So for the most part missing genes is a pretty minimal problem. We will switch to using counts rather than tpm for the next section to test if there are sufficient genes in each cell.

```{r}
#testing for cells with genes under 4000

test2 <- sapply(jeager2018_counts[,c(1:360)], function(y) sum(length(which(y>0))))

# which(test2<4000)
# X151207_E12_N_C_F_EE_2 X151207_H12_N_C_F_EE_2  X151214_F1_P_C_N_EE_3  X151214_H1_P_C_N_EE_3 X151214_H11_P_C_F_EE_3             X160107_07 
#                     49                     83                    123                    137                    139                    199 
#             X160107_08             X160107_22             X160107_23             X160107_29             X160118_09             X160118_14 
#                    200                    214                    215                    221                    249                    254 
#             X160118_16             X160118_17             X160118_21             X160118_27             X160118_28             X160118_29 
#                    256                    257                    261                    267                    268                    269 
#             X160118_30             X160118_31 
#                    270                    271 

# > length(which(test2>3999))
# [1] 340
filtered.idx <- as.numeric(which(test2>4000))
filtered.idx <- order(c(filtered.idx,194))

print("Average gene's passing threshold per cell:")
print(meangenecounts(jeager2018_tpm[,filtered.idx])) 
print("Total missing entries in first dataset:")
print(sum(is.na(jeager2018_tpm[,filtered.idx]))) 
```

So that leaves us with 340 cells, which is a believable number considering we should be eliminating 44 cells eventually. For some reason after the 146th column the naming convention of the cells changes and they stop including the stains they used to identify cell type in the column name.  It is still present in the metadata (where each nuclei is a row).  After 360 all are DG excitatory nuclei.  We are still not getting the same number of average genes passing threshold per cell but we will just have to make do.

Reproducing quick check for batch effects and cell type clustering, supplementary figure 2b and 2c respectfully.

making PCA plots for batch effects: https://tem11010.github.io/Plotting-PCAs/



```{r}
filteredDat <- jeager2018_tpm[,filtered.idx]
filteredDat[is.na(filteredDat)] = 0 
filteredDat <- t(filteredDat)
filteredDat <- data.frame(filteredDat)
filteredDat$batch <- as.factor(jeager2018_meta$Mouse_Number[filtered.idx])
filteredDat$fos_status <- as.factor(sapply(as.character(jeager2018_meta$source_name[filtered.idx]), function(y) if (grepl("_F_", y, fixed=TRUE)) "Fos+" else "Fos-"  ))

#using FactoMineR PCA function
pcs <- PCA(filteredDat, 
           quali.sup=c(40222, 40223), #to identify batch as a qualitative variable
           graph = FALSE)

#putting pcs into our dataframe
filteredDat$pc1 <- pcs$ind$coord[, 1] 

filteredDat$pc2 <- pcs$ind$coord[, 2]
```

```{r}
#calls pcs object to get variences rounds them to 2nd digit makes string for axis labels
axislabs <-c(paste("PC1 (", as.character(round(pcs$eig[1,2],2)),"% of varience)", sep =''),
             paste("PC2 (", as.character(round(pcs$eig[2,2],2)),"% of varience)", sep =''))

p <- ggplot(data = filteredDat, aes(x = pc1, y = pc2, color = batch, shape = fos_status)) +
  geom_point()  + 
  labs(title='First Two PCs of HC and NE neurons from Jeager et al., (2018)',
       color='Batch (Mouse)',
       shape='Fos Status') +
  xlab(axislabs[1]) +
  ylab(axislabs[2]) +
  theme(plot.title = element_text(hjust = 0.5))
```
```{r}
p
```
    Above we have replicated supplemental figure 2b.  I have included the cells that passsed threshold (indexed in filtered.idx) from homecage and 1 hour post novel environment exposure from the first dataset (rows 1-360 of our combined dataframes jeager2018_meta, jeager2018_counts, and jeager2018_tpm).  As the authors stated there are minimal batch effects.

  Many cell types are represented from many hippocampal regions and even a few subiculum and thalamus cells, mostly excitatory cells but there are also some inhibitory cells.  Within this diverse sample the recently active cells, our engram cells are not linearly seperable though in other reports focusing on one particular cell type PCA has been sufficient to separate and cluster engram and non-engram cells.  Rao-Ruiz et al., (2019) did this with dentate gyrus granule cells (DGCs).  How ever here Jeager et al. (2018) used tSNE to find a subspace that separates fos-status across cell types.

Then we can use this guide to do it for the tSNE: https://datavizpyr.com/how-to-make-tsne-plot-in-r/

#library(Rtsne)  for tSNE
```{r}
# library(dplyr)
# filteredDat$celltype <- as.factor(as.character(jeager2018_meta$predicted_cell_type[filtered.idx]))
# 
# set.seed(223)
# tSNE_fit <- filteredDat[, c(1:40223,20226)] %>%
#   as_tibble() %>%
#   select(where(is.numeric)) %>%
#   column_to_rownames("ID") %>%
#   scale() %>% 
#   Rtsne()
# 
# tSNE_fit <- filteredDat[, c(1:40221)] %>%
#   scale() %>% 
#   Rtsne()
```
# move on to the classifier or DGE go to sleep.

 - weird error when I reun the second attempt at tSNE
#Error in na.fail.default(X) : missing values in object
-Not really sure how to correct this, filteredDat has already been filtered for 
0's.  Prehaps I need to transpose the dataframe?


DEG for monocle.  Figure 2c shows that within differetn cell types the DG shows the greatest number of fos+ DEGs relative to the VIP neurons and CA1 prinicple neurons.  Fos- but Arc+ CA1 cells were expluced from the analysis, though the threshold for the exclusion is not given.
```{r}
arc_expr = which(colnames(filteredDat)=="Arc")
fos_expr = which(colnames(filteredDat)=="Fos")
print(paste("Arc expression is on column ",arc_expr))
print(paste("Fos expression is on column ",fos_expr))
```
```{r}
hist(filteredDat[filteredDat$celltype=="CA1", arc_expr])
```

```{r}
ca1fosvarc <- filteredDat[filteredDat$celltype=="CA1", c(arc_expr,fos_expr,40223)]

p <- ggplot(ca1fosvarc, aes(x = Arc, y = Fos, color = fos_status)) +
  geom_point() +
  ggtitle("Arc vs Fos expression and status in CA1") +
  labs(color = 'Fos Status') +
  theme(plot.title = element_text(hjust = 0.5))
```

```{r}
p
```


FILTERING AS PER THE METHODS
---





THE CLASSIFIER IN SIGNATURE ESTABLISHED OVER 4H SELECT REACTIVATED NEURONS
---

From Jeager et al. (2018):  "A receiver-operating characteristics (ROC) curve showed that the model successfully called Reactivated neurons with an area under the curve of 0.93 and 0.96 when compared to either the Newly Activated (model i) or Not Reactivated (model ii) nuclei, respectively (Fig. 6c), and when the model was tasked with distinguishing all three groups simultaneously, classification errors of 13%, 20%, and 20% were calculated for Reactivated, Not Reactivated, and Newly Activated nuclei, respectively."

Notes on which nuclei to include:  "DEGs between Reactivated and Not Reactivated nuclei (FDR < 0.05) were filtered to maintain only those with expression in at least 80% of Reactivated nuclei, for upregulated genes, or at most 40% of Reactivated nuclei for downregulated genes where the cutoff for expression was placed at 1 log2(TPM+1). Genes were further excluded if they contained any association with batch effect or were differentially expressed in the 1-h vs HC comparison. The remaining 191 genes were passed through an initial round of feature elimination. A pairwise classification between Reactivated and either Newly Activated or Not Reactivated nuclei used a random forest classifier with a binomial family and 10,000 trees (R statistical package randomForest78. ROC curves were calculated by training on 15 nuclei per condition and testing on the remaining hold out samples, then calculating sensitivity and specificity using the R statistical package ROCR81. The genes with the top importance (Mean Decrease Gini > 0.4) in each of the pairwise comparisons were then pooled and the random forest procedure was used with a binomial family and 10,000 trees to classify all three conditions together. The classification error rate of this model was reported. This model was then used to predict the presence of a Reactivated gene signature using the 4- and 5-h time points using the base R stats predict function."

-it is not described what an association with batch effect means



References

Jaeger, B. N., Linker, S. B., Parylak, S. L., Barron, J. J., Gallina, I. S., Saavedra, C. D., ... & Gage, F. H. (2018). A novel environment-evoked transcriptional signature predicts reactivity in single dentate granule neurons. Nature communications, 9(1), 1-15.

Rao-Ruiz, P., Couey, J. J., Marcelo, I. M., Bouwkamp, C. G., Slump, D. E., Matos, M. R., ... & Kushner, S. A. (2019). Engram-specific transcriptome profiling of contextual memory consolidation. Nature communications, 10(1), 1-14.
